<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Biological Entity Dictionary (BED) • BED</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Biological Entity Dictionary (BED)">
<meta name="description" content="An interface for the Neo4j database providing mapping between different identifiers of biological entities. This Biological Entity Dictionary (BED) has been developed to address three main challenges. The first one is related to the completeness of identifier mappings. Indeed, direct mapping information provided by the different systems are not always complete and can be enriched by mappings provided by other resources. More interestingly, direct mappings not identified by any of these resources can be indirectly inferred by using mappings to a third reference. For example, many human Ensembl gene ID are not directly mapped to any Entrez gene ID but such mappings can be inferred using respective mappings to HGNC ID. The second challenge is related to the mapping of deprecated identifiers. Indeed, entity identifiers can change from one resource release to another. The identifier history is provided by some resources, such as Ensembl or the NCBI, but it is generally not used by mapping tools. The third challenge is related to the automation of the mapping process according to the relationships between the biological entities of interest. Indeed, mapping between gene and protein ID scopes should not be done the same way than between two scopes regarding gene ID. Also, converting identifiers from different organisms should be possible using gene orthologs information. The method has been published by Godard and van Eyll (2018) &lt;doi:10.12688/f1000research.13925.3&gt;.">
<meta property="og:description" content="An interface for the Neo4j database providing mapping between different identifiers of biological entities. This Biological Entity Dictionary (BED) has been developed to address three main challenges. The first one is related to the completeness of identifier mappings. Indeed, direct mapping information provided by the different systems are not always complete and can be enriched by mappings provided by other resources. More interestingly, direct mappings not identified by any of these resources can be indirectly inferred by using mappings to a third reference. For example, many human Ensembl gene ID are not directly mapped to any Entrez gene ID but such mappings can be inferred using respective mappings to HGNC ID. The second challenge is related to the mapping of deprecated identifiers. Indeed, entity identifiers can change from one resource release to another. The identifier history is provided by some resources, such as Ensembl or the NCBI, but it is generally not used by mapping tools. The third challenge is related to the automation of the mapping process according to the relationships between the biological entities of interest. Indeed, mapping between gene and protein ID scopes should not be done the same way than between two scopes regarding gene ID. Also, converting identifiers from different organisms should be possible using gene orthologs information. The method has been published by Godard and van Eyll (2018) &lt;doi:10.12688/f1000research.13925.3&gt;.">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-7RG3ZVT8WZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-7RG3ZVT8WZ');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">BED</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/BED.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/patzaw/BED/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="readme">README<a class="anchor" aria-label="anchor" href="#readme"></a>
</h1></div>
</div>
<div class="section level1">
<h1 id="biological-entity-dictionary-">Biological Entity Dictionary <img src="reference/figures/BED.png" align="right" alt="" width="120"><a class="anchor" aria-label="anchor" href="#biological-entity-dictionary-"></a>
</h1>
<p><a href="https://cran.r-project.org/package=BED" class="external-link"><img src="https://www.r-pkg.org/badges/version/BED" alt="CRAN_Status_Badge"></a> <a href="https://cran.r-project.org/package=BED" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/BED"></a></p>
<p>An interface for the ‘Neo4j’ database providing mapping between different identifiers of biological entities. This Biological Entity Dictionary (BED) has been developed to address three main challenges. The first one is related to the completeness of identifier mappings. Indeed, direct mapping information provided by the different systems are not always complete and can be enriched by mappings provided by other resources. More interestingly, direct mappings not identified by any of these resources can be indirectly inferred by using mappings to a third reference. For example, many human Ensembl gene ID are not directly mapped to any Entrez gene ID but such mappings can be inferred using respective mappings to HGNC ID. The second challenge is related to the mapping of deprecated identifiers. Indeed, entity identifiers can change from one resource release to another. The identifier history is provided by some resources, such as Ensembl or the NCBI, but it is generally not used by mapping tools. The third challenge is related to the automation of the mapping process according to the relationships between the biological entities of interest. Indeed, mapping between gene and protein ID scopes should not be done the same way than between two scopes regarding gene ID. Also, converting identifiers from different organisms should be possible using gene orthologs information. The method has been published by Godard and van Eyll (2018) <a href="doi:10.12688/f1000research.13925.3" class="uri">doi:10.12688/f1000research.13925.3</a>.</p>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="from-cran">From CRAN<a class="anchor" aria-label="anchor" href="#from-cran"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BED"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<p>The following R packages available on CRAN are required:</p>
<ul>
<li>
<a href="https://CRAN.R-project.org/package=neo2R" class="external-link">neo2R</a>: Neo4j to R</li>
<li>
<a href="https://CRAN.R-project.org/package=visNetwork" class="external-link">visNetwork</a>: Network Visualization using ‘vis.js’ Library</li>
<li>
<a href="https://CRAN.R-project.org/package=dplyr" class="external-link">dplyr</a>: A Grammar of Data Manipulation</li>
<li>
<a href="https://CRAN.R-project.org/package=readr" class="external-link">readr</a>: Read Rectangular Text Data</li>
<li>
<a href="https://CRAN.R-project.org/package=stringr" class="external-link">stringr</a>: Simple, Consistent Wrappers for Common String Operations</li>
<li>
<a href="https://CRAN.R-project.org/package=utils" class="external-link">utils</a>: The R Utils Package</li>
<li>
<a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a>: Web Application Framework for R</li>
<li>
<a href="https://CRAN.R-project.org/package=DT" class="external-link">DT</a>: A Wrapper of the JavaScript Library ‘DataTables’</li>
<li>
<a href="https://CRAN.R-project.org/package=miniUI" class="external-link">miniUI</a>: Shiny UI Widgets for Small Screens</li>
<li>
<a href="https://CRAN.R-project.org/package=rstudioapi" class="external-link">rstudioapi</a>: Safely Access the RStudio API</li>
</ul>
<p>And those are suggested:</p>
<ul>
<li>
<a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a>: A General-Purpose Package for Dynamic Report Generation in R</li>
<li>
<a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a>: Dynamic Documents for R</li>
<li>
<a href="https://CRAN.R-project.org/package=biomaRt" class="external-link">biomaRt</a>: Interface to BioMart databases (i.e. Ensembl)</li>
<li>
<a href="https://CRAN.R-project.org/package=GEOquery" class="external-link">GEOquery</a>: Get data from NCBI Gene Expression Omnibus (GEO)</li>
<li>
<a href="https://CRAN.R-project.org/package=base64enc" class="external-link">base64enc</a>: Tools for base64 encoding</li>
<li>
<a href="https://CRAN.R-project.org/package=htmltools" class="external-link">htmltools</a>: Tools for HTML</li>
<li>
<a href="https://CRAN.R-project.org/package=webshot2" class="external-link">webshot2</a>: Take Screenshots of Web Pages</li>
<li>
<a href="https://CRAN.R-project.org/package=RCurl" class="external-link">RCurl</a>: General Network (HTTP/FTP/…) Client Interface for R</li>
</ul>
</div>
<div class="section level3">
<h3 id="installation-from-github">Installation from github<a class="anchor" aria-label="anchor" href="#installation-from-github"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"patzaw/BED"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="possible-issue-when-updating-from-releases--130">Possible issue when updating from releases &lt;= 1.3.0<a class="anchor" aria-label="anchor" href="#possible-issue-when-updating-from-releases--130"></a>
</h3>
<p>If you get an error like the following…</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>Error<span class="sc">:</span> package or namespace load failed <span class="cf">for</span> ‘BED’<span class="sc">:</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a> .onLoad failed <span class="cf">in</span> <span class="fu">loadNamespace</span>() <span class="cf">for</span> <span class="st">'BED'</span>, details<span class="sc">:</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  call<span class="sc">:</span> connections[[connection]][[<span class="st">"cache"</span>]]</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  error<span class="sc">:</span> subscript out of bounds</span></code></pre></div>
<p>… remove the BED folder located here:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HOME"</span><span class="op">)</span>, <span class="st">"R"</span>, <span class="st">"BED"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>Documentation is provided in the <a href="https://patzaw.github.io/BED/articles/BED.html">BED</a> vignette.</p>
<p>A public instance of the <a href="#available-bed-database-instance">BED Neo4j database</a> is provided for convenience and can be reached as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patzaw.github.io/BED/">BED</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/connectToBed.html">connectToBed</a></span><span class="op">(</span><span class="st">"https://genodesy.org/BED/"</span>, remember<span class="op">=</span><span class="cn">TRUE</span>, useCache<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/findBeids.html">findBeids</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
<div class="section level2">
<h2 id="citing-bed">Citing BED<a class="anchor" aria-label="anchor" href="#citing-bed"></a>
</h2>
<p>This package and the underlying research has been published in this peer reviewed article:</p>
<p><a href="https://doi.org/10.12688/f1000research.13925.3" target="_blank" class="external-link"> Patrice Godard and Jonathan van Eyll (2018). BED: a Biological Entity Dictionary based on a graph data model (version 3; peer review: 2 approved). F1000Research, 7:195. </a></p>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
<div class="section level2">
<h2 id="available-bed-database-instance">Available BED database instance<a class="anchor" aria-label="anchor" href="#available-bed-database-instance"></a>
</h2>
<p>An instance of the BED database (UCB-Human) has been built using the script provided in the BED R package.</p>
<p>This instance is focused on <em>Homo sapiens</em>, <em>Mus musculus</em>, <em>Rattus norvegicus</em>, <em>Sus scrofa</em> and <em>Danio rerio</em> organisms. It has been built from the following resources:</p>
<ul>
<li>Ensembl</li>
<li>NCBI</li>
<li>Uniprot</li>
<li>biomaRt</li>
<li>GEOquery</li>
</ul>
<p>The Neo4j graph database is available as a dump file shared in <a href="https://zenodo.org/records/14048653" class="external-link">Zenodo</a>.</p>
<p>The following shell commands can be adapted according to user needs and called to get a running Neo4j container with a BED database instance.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co">#!/bin/sh</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">####################################################@</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">## Check folders ----</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="cf">if</span> <span class="bu">test</span> <span class="at">-e</span> ~/.cache/BED/neo4jData<span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>   <span class="bu">echo</span> <span class="st">"~/.cache/BED/neo4jData directory exists ==&gt; abort - Remove it before proceeding"</span> <span class="op">&gt;&amp;</span><span class="dv">2</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>   <span class="bu">exit</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/.cache/BED/neo4jData</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="cf">if</span> <span class="bu">test</span> <span class="at">-e</span> ~/.cache/BED/neo4jDump<span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>   <span class="bu">echo</span> <span class="st">"~/.cache/BED/neo4jDump directory exists ==&gt; abort - Remove it before proceeding"</span> <span class="op">&gt;&amp;</span><span class="dv">2</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>   <span class="bu">exit</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/.cache/BED/neo4jDump</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">####################################################@</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">## Download data ----</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="bu">export</span> <span class="va">BED_REP_URL</span><span class="op">=</span>https://zenodo.org/records/14048653/files/</span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="fu">wget</span> <span class="va">$BED_REP_URL</span>/dump_bed_Genodesy-Human_2024.11.07.dump <span class="at">-O</span> ~/.cache/BED/neo4jDump/neo4j.dump</span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">####################################################@</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">## Import data ----</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--interactive</span> <span class="at">--tty</span> <span class="at">--rm</span> <span class="dt">\</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a>   <span class="at">--volume</span><span class="op">=</span>~/.cache/BED/neo4jData/data:/data <span class="dt">\</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a>   <span class="at">--volume</span><span class="op">=</span>~/.cache/BED/neo4jDump:/backups <span class="dt">\</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a>    neo4j:5.25.1 <span class="dt">\</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a>neo4j-admin database load neo4j <span class="at">--from-path</span><span class="op">=</span>/backups</span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">####################################################@</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">## Start neo4j ----</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-d</span> <span class="dt">\</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a>   <span class="at">--name</span> bed_2024.11.07 <span class="dt">\</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a>   <span class="at">--publish</span><span class="op">=</span>5454:7474 <span class="dt">\</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a>   <span class="at">--publish</span><span class="op">=</span>5687:7687 <span class="dt">\</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a>   <span class="at">--env</span><span class="op">=</span>NEO4J_dbms_memory_heap_initial__size=4G <span class="dt">\</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a>   <span class="at">--env</span><span class="op">=</span>NEO4J_dbms_memory_heap_max__size=4G <span class="dt">\</span></span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a>   <span class="at">--env</span><span class="op">=</span>NEO4J_dbms_memory_pagecache_size=4G <span class="dt">\</span></span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a>   <span class="at">--env</span><span class="op">=</span>NEO4J_dbms_read__only=true <span class="dt">\</span></span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a>   <span class="at">--env</span><span class="op">=</span>NEO4J_AUTH=none <span class="dt">\</span></span>
<span id="cb6-41"><a href="#cb6-41" tabindex="-1"></a>   <span class="at">--volume</span> ~/.cache/BED/neo4jData/data:/data <span class="dt">\</span></span>
<span id="cb6-42"><a href="#cb6-42" tabindex="-1"></a>   <span class="at">--volume</span> ~/.cache/BED/neo4jData/logs:/var/lib/neo4j/logs <span class="dt">\</span></span>
<span id="cb6-43"><a href="#cb6-43" tabindex="-1"></a>   <span class="at">--restart</span><span class="op">=</span>always <span class="dt">\</span></span>
<span id="cb6-44"><a href="#cb6-44" tabindex="-1"></a>   neo4j:5.25.1</span></code></pre></div>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
<div class="section level2">
<h2 id="build-a-bed-database-instance">Build a BED database instance<a class="anchor" aria-label="anchor" href="#build-a-bed-database-instance"></a>
</h2>
<p>Building and feeding a BED database instance is achieved using scripts available in the “supp/Build” folder.</p>
<div class="section level3">
<h3 id="run-a-neo4j-docker-images">Run a neo4j docker images<a class="anchor" aria-label="anchor" href="#run-a-neo4j-docker-images"></a>
</h3>
<p>Using the S01-NewBED-Container.sh script.</p>
</div>
<div class="section level3">
<h3 id="build-and-feed-bed">Build and feed BED<a class="anchor" aria-label="anchor" href="#build-and-feed-bed"></a>
</h3>
<p>Using the S02-Rebuild-BED.sh script which compile the Rebuild-BED.Rmd document.</p>
</div>
</div>
<div class="section level2">
<h2 id="dump-the-graph-db-content-for-sharing">Dump the graph DB content for sharing<a class="anchor" aria-label="anchor" href="#dump-the-graph-db-content-for-sharing"></a>
</h2>
<p>Using the S03-Dump-BED.sh script</p>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
<div class="section level2">
<h2 id="docker-notes">Docker notes<a class="anchor" aria-label="anchor" href="#docker-notes"></a>
</h2>
<p><a href="https://github.com/sespesogil" class="external-link">Sergio Espeso-Gil</a> has reported stability issues with Docker images in Windows. It’s mainly solved by checking the “Use the WSL2 based engine” options in docker settings. More information is provided here: <a href="https://docs.docker.com/docker-for-windows/wsl/" class="external-link uri">https://docs.docker.com/docker-for-windows/wsl/</a></p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=BED" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/patzaw/BED/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/patzaw/BED/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing BED</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://patzaw.github.io" class="external-link">Patrice Godard</a> <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0001-6257-9730" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://patzaw.github.io" class="external-link">Patrice Godard</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
