<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BED">
<title>Biological Entity Dictionary (BED): exploring and converting identifiers of biological entities such as genes, transcripts or peptides • BED</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Biological Entity Dictionary (BED): exploring and converting identifiers of biological entities such as genes, transcripts or peptides">
<meta property="og:description" content="BED">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-7RG3ZVT8WZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-7RG3ZVT8WZ');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BED</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/BED.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/patzaw/BED/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="BED_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="BED_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="BED_files/vis-9.1.0/vis-network.min.css" rel="stylesheet">
<script src="BED_files/vis-9.1.0/vis-network.min.js"></script><script src="BED_files/visNetwork-binding-2.1.2/visNetwork.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Biological Entity Dictionary (BED): exploring and converting identifiers of biological entities such as genes, transcripts or peptides</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/patzaw/BED/blob/HEAD/vignettes/BED.Rmd" class="external-link"><code>vignettes/BED.Rmd</code></a></small>
      <div class="d-none name"><code>BED.Rmd</code></div>
    </div>

    
    
<div style="width:200px;">
<p><img src="img/BED.png" style="width:100.0%"></p>
</div>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This Biological Entity Dictionary (BED) has been developed to address
three main challenges. The first one is related to the completeness of
identifier mappings. Indeed, direct mapping information provided by the
different systems are not always complete and can be enriched by
mappings provided by other resources. More interestingly, direct
mappings not identified by any of these resources can be indirectly
inferred by using mappings to a third reference. For example, many human
Ensembl gene ID are not directly mapped to any Entrez gene ID but such
mappings can be inferred using respective mappings to HGNC ID. The
second challenge is related to the mapping of deprecated identifiers.
Indeed, entity identifiers can change from one resource release to
another. The identifier history is provided by some resources, such as
Ensembl or the NCBI, but it is generally not used by mapping tools. The
third challenge is related to the automation of the mapping process
according to the relationships between the biological entities of
interest. Indeed, mapping between gene and protein ID scopes should not
be done the same way than between two scopes regarding gene ID. Also,
converting identifiers from different organisms should be possible using
gene orthologs information.</p>
<p>This document shows how to use the <strong>BED (Biological Entity
Dictionary)</strong> R package to get and explore mapping between
identifiers of biological entities (BE). This package provides a way to
connect to a BED Neo4j database in which the relationships between the
identifiers from different sources are recorded.</p>
<div class="section level3">
<h3 id="citing-bed">Citing BED<a class="anchor" aria-label="anchor" href="#citing-bed"></a>
</h3>
<p>This package and the underlying research has been published in this
peer reviewed article:</p>
<p><a href="https://doi.org/10.12688/f1000research.13925.3" target="_blank" class="external-link">
Patrice Godard and Jonathan van Eyll (2018). BED: a Biological Entity
Dictionary based on a graph data model (version 3; peer review: 2
approved). F1000Research, 7:195. </a></p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="section level4">
<h4 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h4>
<p>This BED package depends on the following packages available in the
CRAN repository:</p>
<ul>
<li><strong>neo2R</strong></li>
<li><strong>visNetwork</strong></li>
<li><strong>dplyr</strong></li>
<li><strong>readr</strong></li>
<li><strong>stringr</strong></li>
<li><strong>utils</strong></li>
<li><strong>shiny</strong></li>
<li><strong>DT</strong></li>
<li><strong>miniUI</strong></li>
<li><strong>rstudioapi</strong></li>
</ul>
<p>All these packages must be installed before installing BED.</p>
</div>
<div class="section level4">
<h4 id="installation-from-github">Installation from github<a class="anchor" aria-label="anchor" href="#installation-from-github"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"patzaw/BED"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="possible-issue-when-updating-from-releases-1-3-0">Possible issue when updating from releases &lt;= 1.3.0<a class="anchor" aria-label="anchor" href="#possible-issue-when-updating-from-releases-1-3-0"></a>
</h4>
<p>If you get an error like the following…</p>
<pre><code>Error: package or namespace load failed for ‘BED’:
 .onLoad failed in loadNamespace() for 'BED', details:
  call: connections[[connection]][["cache"]]
  error: subscript out of bounds</code></pre>
<p>… remove the BED folder located here:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HOME"</span><span class="op">)</span>, <span class="st">"R"</span>, <span class="st">"BED"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="connection">Connection<a class="anchor" aria-label="anchor" href="#connection"></a>
</h3>
<p>Before using BED, the connection needs to be established with the
underlying Neo4j DB. <code>url</code>, <code>username</code> and
<code>password</code> should be adapted.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patzaw.github.io/BED/">BED</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/connectToBed.html">connectToBed</a></span><span class="op">(</span>url<span class="op">=</span><span class="st">"localhost:5454"</span>, remember<span class="op">=</span><span class="cn">FALSE</span>, useCache<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The <code>remember</code> parameter can be set to <code>TRUE</code>
in order to save connection information that will be automatically used
the next time the <code><a href="../reference/connectToBed.html">connectToBed()</a></code> function is called. By
default, this parameter is set to <code>FALSE</code> to comply with CRAN
policies. Saved connection can be managed with the
<code><a href="../reference/lsBedConnections.html">lsBedConnections()</a></code> and the
<code><a href="../reference/forgetBedConnection.html">forgetBedConnection()</a></code> functions.</p>
<p>The <code>useCache</code> parameter is by default set to
<code>FALSE</code> to comply with CRAN policies. However, it is
recommended to set it to <code>TRUE</code> to improve the speed of
recurrent queries: the results of some large queries are saved locally
in a file.</p>
<p>The connection can be checked the following way.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/checkBedConn.html">checkBedConn</a></span><span class="op">(</span>verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## https://genodesy.org/BED</span></span></code></pre>
<pre><code><span><span class="co">## BED</span></span></code></pre>
<pre><code><span><span class="co">## Genodesy-Human</span></span></code></pre>
<pre><code><span><span class="co">## 2024.05.15</span></span></code></pre>
<pre><code><span><span class="co">## Cache ON</span></span></code></pre>
<pre><code><span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## attr(,"dbVersion")</span></span>
<span><span class="co">##   name       instance    version</span></span>
<span><span class="co">## 1  BED Genodesy-Human 2024.05.15</span></span></code></pre>
<p>If the <code>verbose</code> parameter is set to TRUE, the URL and the
content version are displayed as messages.</p>
<p>The following function list saved connections.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/lsBedConnections.html">lsBedConnections</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The <code>connection</code> param of the <code>connectToBed</code>
function can be used to connect to a saved connection other than the
last one.</p>
</div>
<div class="section level3">
<h3 id="data-model">Data model<a class="anchor" aria-label="anchor" href="#data-model"></a>
</h3>
<p>The BED underlying data model can be shown at any time using the
following command.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/showBedDataModel.html">showBedDataModel</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/BED-data-model.png"></p>
</div>
<div class="section level3">
<h3 id="direct-calls">Direct calls<a class="anchor" aria-label="anchor" href="#direct-calls"></a>
</h3>
<p>Cypher queries can be run directly on the Neo4j database using the
<code>cypher</code> function from the <strong>neo2R</strong> package
through the <code>bedCall</code> function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bedCall.html">bedCall</a></span><span class="op">(</span></span>
<span>    <span class="va">cypher</span>,</span>
<span>    query<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/neo2R/man/prepCql.html" class="external-link">prepCql</a></span><span class="op">(</span></span>
<span>       <span class="st">'MATCH (n:BEID)'</span>,</span>
<span>       <span class="st">'WHERE n.value IN $values'</span>,</span>
<span>       <span class="st">'RETURN DISTINCT n.value AS value, labels(n), n.database'</span></span>
<span>    <span class="op">)</span>,</span>
<span>    parameters<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10"</span>, <span class="st">"100"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">results</span></span></code></pre></div>
<pre><code><span><span class="co">##   value      labels(n) n.database</span></span>
<span><span class="co">## 1    10 BEID || GeneID EntrezGene</span></span>
<span><span class="co">## 2   100 BEID || GeneID EntrezGene</span></span>
<span><span class="co">## 3   100 BEID || GeneID       HGNC</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="feeding-the-database">Feeding the database<a class="anchor" aria-label="anchor" href="#feeding-the-database"></a>
</h3>
<p>Many functions are provided within the package to build your own BED
database instance. These functions are not exported in order to avoid
their use when interacting with BED normally. Information about how to
get an instance of the BED neo4j database is provided here:</p>
<ul>
<li><a href="https://github.com/patzaw/BED#bed-database-instance-available-as-a-docker-image" class="external-link uri">https://github.com/patzaw/BED#bed-database-instance-available-as-a-docker-image</a></li>
<li><a href="https://github.com/patzaw/BED#build-a-bed-database-instance" class="external-link uri">https://github.com/patzaw/BED#build-a-bed-database-instance</a></li>
</ul>
<p>It can be adapted to user needs.</p>
</div>
<div class="section level3">
<h3 id="caching">Caching<a class="anchor" aria-label="anchor" href="#caching"></a>
</h3>
<p>This part is relevant if the <code>useCache</code> parameter is set
to TRUE when calling <code><a href="../reference/connectToBed.html">connectToBed()</a></code>.</p>
<p>Functions of the BED package used to retrieve thousands of
identifiers can take some time (generally a few seconds) before
returning a result. Thus for this kind of query, the query is run for
all the relevant ID in the DB and thanks to a cache system implemented
in the package same queries with different filters should be much faster
the following times.</p>
<p>By default the cache is flushed when the system detect
inconsistencies with the BED database. However, it can also be manualy
flushed if needed using the <code><a href="../reference/clearBedCache.html">clearBedCache()</a></code> function.</p>
<p>Queries already in cache can be listed using the
<code><a href="../reference/lsBedCache.html">lsBedCache()</a></code> function which also return the occupied disk
space.</p>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
</div>
<div class="section level2">
<h2 id="exploring-available-data">Exploring available data<a class="anchor" aria-label="anchor" href="#exploring-available-data"></a>
</h2>
<div class="section level3">
<h3 id="biological-entities">Biological entities<a class="anchor" aria-label="anchor" href="#biological-entities"></a>
</h3>
<p>BED is organized around the central concept of <strong>Biological
Entity</strong> (BE). All supported types of BE can be listed.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listBe.html">listBe</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene"       "Transcript" "Peptide"    "Object"</span></span></code></pre>
<p>These BE are organized according to how they are related to each
other. For example a <em>Gene</em> <em>is_expressed_as</em> a
<em>Transcript</em>. This organization allows to find the first upstream
BE common to a set of BE.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/firstCommonUpstreamBe.html">firstCommonUpstreamBe</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Object"</span>, <span class="st">"Transcript"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene"</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/firstCommonUpstreamBe.html">firstCommonUpstreamBe</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Peptide"</span>, <span class="st">"Transcript"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Transcript"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="organisms">Organisms<a class="anchor" aria-label="anchor" href="#organisms"></a>
</h3>
<p>Several organims can be supported by the BED underlying database.
They can be listed the following way.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listOrganisms.html">listOrganisms</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Danio rerio"       "Homo sapiens"      "Sus scrofa"       </span></span>
<span><span class="co">## [4] "Mus musculus"      "Rattus norvegicus"</span></span></code></pre>
<p>Common names are also supported and the corresponding taxonomic
identifiers can be retrieved. Conversely the organism names
corresponding to a taxonomic ID can be listed.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getOrgNames.html">getOrgNames</a></span><span class="op">(</span><span class="fu"><a href="../reference/getTaxId.html">getTaxId</a></span><span class="op">(</span><span class="st">"human"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   taxID                        name           nameClass</span></span>
<span><span class="co">## 1  9606 Homo sapiens Linnaeus, 1758           authority</span></span>
<span><span class="co">## 2  9606                       human genbank common name</span></span>
<span><span class="co">## 3  9606                Homo sapiens     scientific name</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="identifiers-of-biological-entities">Identifiers of biological entities<a class="anchor" aria-label="anchor" href="#identifiers-of-biological-entities"></a>
</h3>
<p>The main aim of BED is to allow the mapping of identifiers from
different sources such as Ensembl or Entrez. Supported sources can be
listed the following way for each supported organism.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listBeIdSources.html">listBeIdSources</a></span><span class="op">(</span>be<span class="op">=</span><span class="st">"Transcript"</span>, organism<span class="op">=</span><span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             database   nbBe   nbId         be</span></span>
<span><span class="co">## 1 BEDTech_transcript 114302 114302 Transcript</span></span>
<span><span class="co">## 2             RefSeq 270808 279118 Transcript</span></span>
<span><span class="co">## 3     Ens_transcript 277147 288548 Transcript</span></span></code></pre>
<p>The database gathering the largest number of BE of specific type can
also be identified.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/largestBeSource.html">largestBeSource</a></span><span class="op">(</span>be<span class="op">=</span><span class="st">"Transcript"</span>, organism<span class="op">=</span><span class="st">"human"</span>, restricted<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Ens_transcript"</span></span></code></pre>
<p>Finally, the <code><a href="../reference/getAllBeIdSources.html">getAllBeIdSources()</a></code> function returns all
the source databases of BE identifiers whatever the BE type.</p>
</div>
<div class="section level3">
<h3 id="experimental-platforms-and-probes">Experimental platforms and probes<a class="anchor" aria-label="anchor" href="#experimental-platforms-and-probes"></a>
</h3>
<p>BED also supports experimental platforms and provides mapping
betweens probes and BE identifiers (BEID).</p>
<p>The supported platforms can be listed the following way. The
<code><a href="../reference/getTargetedBe.html">getTargetedBe()</a></code> function returns the type of BE on which a
specific platform focus.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/listPlatforms.html">listPlatforms</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              name                                        description focus</span></span>
<span><span class="co">## GPL6101   GPL6101        Illumina ratRef-12 v1.0 expression beadchip  Gene</span></span>
<span><span class="co">## GPL6887   GPL6887        Illumina MouseWG-6 v2.0 expression beadchip  Gene</span></span>
<span><span class="co">## GPL6947   GPL6947       Illumina HumanHT-12 V3.0 expression beadchip  Gene</span></span>
<span><span class="co">## GPL10558 GPL10558       Illumina HumanHT-12 V4.0 expression beadchip  Gene</span></span>
<span><span class="co">## GPL1355   GPL1355     [Rat230_2] Affymetrix Rat Genome 230 2.0 Array  Gene</span></span>
<span><span class="co">## GPL1261   GPL1261 [Mouse430_2] Affymetrix Mouse Genome 430 2.0 Array  Gene</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getTargetedBe.html">getTargetedBe</a></span><span class="op">(</span><span class="st">"GPL570"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene"</span></span></code></pre>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
</div>
<div class="section level2">
<h2 id="managing-identifiers">Managing identifiers<a class="anchor" aria-label="anchor" href="#managing-identifiers"></a>
</h2>
<div class="section level3">
<h3 id="retrieving-all-identifiers-from-a-source">Retrieving all identifiers from a source<a class="anchor" aria-label="anchor" href="#retrieving-all-identifiers-from-a-source"></a>
</h3>
<p>All identifiers of an organism BEs from one source can be
retrieved.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBeIds.html">getBeIds</a></span><span class="op">(</span></span>
<span>    be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"human"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">beids</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 216039      5</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">beids</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     id preferred   Gene db.version db.deprecated</span></span>
<span><span class="co">## 1 4535     FALSE 869296   20240516         FALSE</span></span>
<span><span class="co">## 2 4536     FALSE 869300   20240516         FALSE</span></span>
<span><span class="co">## 3 4512     FALSE 869306   20240516         FALSE</span></span>
<span><span class="co">## 4 4513     FALSE 869309   20240516         FALSE</span></span>
<span><span class="co">## 5 4509     FALSE 869311   20240516         FALSE</span></span>
<span><span class="co">## 6 4508     FALSE 869312   20240516         FALSE</span></span></code></pre>
<p>The first column, <em>id</em>, corresponds to the identifiers of the
BE in the source. The column named according to the BE type (in this
case <em>Gene</em>) corresponds to the internal identifier of the
related BE. <strong>BE CAREFUL, THIS INTERNAL ID IS NOT STABLE AND
CANNOT BE USED AS A REFERENCE</strong>. This internal identifier is
useful to identify BEIDS corresponding to the same BE. The following
code can be used to have an overview of such redundancy.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">beids</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##      1      2      3      4      5      6      7      8      9     10     11 </span></span>
<span><span class="co">## 178544   9711   3152   1026    400    154     89     46     24     11      9 </span></span>
<span><span class="co">##     12     13     14     16     31 </span></span>
<span><span class="co">##      6      2      2      1      1</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ambId</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">beids</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">beids</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span><span class="op">&gt;=</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In the example above we can see that most of Gene BE are identified
by only one EntrezGene ID. However many of them are identified by two or
more ID; 32 BE are even identified by 10 or more EntrezGeneID. In this
case, most of these redundancies come from ID history extracted from
Entrez. Legacy ID can be excluded from the retrieved ID by setting the
<code>restricted</code> parameter to TRUE.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBeIds.html">getBeIds</a></span><span class="op">(</span></span>
<span>    be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"human"</span>,</span>
<span>    restricted <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">beids</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 193382      5</span></span></code></pre>
<p>The same code as above can be used to identify remaining
redundancies.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">beids</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##      1      2      3 </span></span>
<span><span class="co">## 192979    194      5</span></span></code></pre>
<p>In the example above we can see that allmost all Gene BE are
identified by only one EntrezGene ID. However some of them are
identified by two or more ID. This result comes from how the BED
database is constructed according to the ID mapping provided by the
different source databases. The graph below shows how the mapping was
done for such a BE with redundant EntrezGene IDs.</p>
<p><strong> This issue has been mainly solved by not taking into account
ambigous mappings between NCBI Entrez gene identifiers and Ensembl gene
identifier provided by Ensembl. It has been achieved using the
<code><a href="../reference/cleanDubiousXRef.html">cleanDubiousXRef()</a></code> function from the 2019.10.11 version of
the BED-UCB-Human database. </strong></p>
<div style="border:solid;overflow:hidden;">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eid</span> <span class="op">&lt;-</span> <span class="va">beids</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">beids</span><span class="op">$</span><span class="va">Gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">beids</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span><span class="op">&gt;=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">eid</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "101927189"</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/exploreBe.html">exploreBe</a></span><span class="op">(</span>id<span class="op">=</span><span class="va">eid</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, be<span class="op">=</span><span class="st">"Gene"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>   <span class="fu"><a href="https://rdrr.io/pkg/visNetwork/man/visPhysics.html" class="external-link">visPhysics</a></span><span class="op">(</span>solver<span class="op">=</span><span class="st">"repulsion"</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"nodes":{"id":[5451248,961246,5450618,629303,857403,6104846,1012188,1229573,1229572,5777745,5777466,1049103],"type":["GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","BESymbol","BESymbol","BESymbol"],"label":["101927189","21355","100874282","90523","ENSG00000146147","41461","614106","ENSG00000236996","ENSG00000236740","LOC101927189","MLIP-IT1","MLIP"],"database":["EntrezGene","HGNC","EntrezGene","EntrezGene","Ens_gene","HGNC","MIM_GENE","Ens_gene","Ens_gene","","",""],"platform":["","","","","","","","","","","",""],"preferred":[true,false,true,true,true,false,false,false,false,null,null,null],"url":["https://www.ncbi.nlm.nih.gov/gene/101927189","http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=21355","https://www.ncbi.nlm.nih.gov/gene/100874282","https://www.ncbi.nlm.nih.gov/gene/90523","http://www.ensembl.org/id/ENSG00000146147","http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=41461","http://www.omim.org/entry/614106","http://www.ensembl.org/id/ENSG00000236996","http://www.ensembl.org/id/ENSG00000236740","","",""],"source":["EntrezGene","HGNC","EntrezGene","EntrezGene","Ens_gene","HGNC","MIM_GENE","Ens_gene","Ens_gene","","",""],"title":["<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/101927189\" target=\"_blank\">101927189<\/a><br><emph>EntrezGene<\/emph><br>Preferred<\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=21355\" target=\"_blank\">21355<\/a><br><emph>HGNC<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/100874282\" target=\"_blank\">100874282<\/a><br><emph>EntrezGene<\/emph><br>Preferred<\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/90523\" target=\"_blank\">90523<\/a><br><emph>EntrezGene<\/emph><br>Preferred<\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/ENSG00000146147\" target=\"_blank\">ENSG00000146147<\/a><br><emph>Ens_gene<\/emph><br>Preferred<\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=41461\" target=\"_blank\">41461<\/a><br><emph>HGNC<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.omim.org/entry/614106\" target=\"_blank\">614106<\/a><br><emph>MIM_GENE<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/ENSG00000236996\" target=\"_blank\">ENSG00000236996<\/a><br><emph>Ens_gene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/ENSG00000236740\" target=\"_blank\">ENSG00000236740<\/a><br><emph>Ens_gene<\/emph><\/p>","<p><strong>BESymbol<\/strong><br>LOC101927189<\/p>","<p><strong>BESymbol<\/strong><br>MLIP-IT1<\/p>","<p><strong>BESymbol<\/strong><br>MLIP<\/p>"],"shape":["dot","dot","dot","dot","dot","dot","dot","dot","dot","box","box","box"],"color.background":["#67E781","#E2DBE6","#67E781","#67E781","#7E58DB","#E2DBE6","#B694E4","#7E58DB","#7E58DB","#E4AC77","#E4AC77","#E4AC77"],"borderWidth":[4,1,4,4,4,1,1,1,1,1,1,1],"borderWidthSelected":[4,2,4,4,4,2,2,2,2,2,2,2],"color.border":["black","black","black","black","black","black","black","black","black","black","black","black"]},"edges":{"id":[1873980,2033171,13405723,13386654,3727076,2226431,2036838,2195125,2581570,2581569,12946182,12945327,12923414,2355858],"title":["corresponds_to","corresponds_to","corresponds_to","corresponds_to","corresponds_to","corresponds_to","is_associated_to","is_associated_to","is_replaced_by","is_replaced_by","is_known_as","is_known_as","is_known_as","is_known_as"],"from":[5451248,5450618,5450618,629303,629303,857403,1012188,1012188,1229573,1229572,5451248,5450618,629303,857403],"to":[857403,857403,6104846,961246,857403,961246,629303,857403,857403,857403,5777745,5777466,1049103,1049103],"arrows":["to","to","to","to","to","to","to","to","to","to","to","to","to","to"],"dashes":[false,false,false,false,false,false,true,true,true,true,false,false,false,false]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot"},"manipulation":{"enabled":false},"interaction":{"selectable":true,"zoomSpeed":1},"physics":{"solver":"repulsion"}},"groups":null,"width":null,"height":null,"idselection":{"enabled":true,"style":"width: 150px; height: 26px","useLabels":true,"main":"Select by id","selected":5451248},"byselection":{"enabled":false,"style":"width: 150px; height: 26px","multiple":false,"hideColor":"rgba(200,200,200,0.5)","highlight":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);","highlight":{"enabled":true,"hoverNearest":false,"degree":1,"algorithm":"all","hideColor":"rgba(200,200,200,0.5)","labelOnly":true},"collapse":{"enabled":false,"fit":false,"resetHighlight":true,"clusterOptions":null,"keepCoord":true,"labelSuffix":"(cluster)"},"legend":{"width":0.2,"useGroups":false,"position":"left","ncol":2,"stepX":100,"stepY":100,"zoom":true,"nodes":[{"label":"GeneID","shape":"dot","color":null},{"label":"BESymbol","shape":"box","color":"#E4AC77"},{"label":"EntrezGene","shape":"dot","color":"#67E781"},{"label":"HGNC","shape":"dot","color":"#E2DBE6"},{"label":"Ens_gene","shape":"dot","color":"#7E58DB"},{"label":"MIM_GENE","shape":"dot","color":"#B694E4"}],"nodesToDataframe":false}},"evals":[],"jsHooks":[]}</script>
</div>
<p><br></p>
<p>The way the ID correspondances are reported in the different source
databases leads to this mapping ambiguity which has to be taken into
account when comparing identifiers from different databases.</p>
<p>The <code><a href="../reference/getBeIds.html">getBeIds()</a></code> returns other columns providing
additional information about the <em>id</em>. The same function can be
used to retrieved symbols or probe identifiers.</p>
<div class="section level4">
<h4 id="preferred-identifier">Preferred identifier<a class="anchor" aria-label="anchor" href="#preferred-identifier"></a>
</h4>
<p>The BED database is constructed according to the relationships
between identifiers provided by the different sources. Biological
entities (BE) are identified as clusters of identifiers which correspond
to each other directly or indirectly (<code>corresponds_to</code>
relationship). Because of this design a BE can be identified by multiple
identifiers (BEID) from the same database as shown above. These BEID are
often related to alternate version of an entity.</p>
<p>For example, Ensembl provides different version (alternative
sequences) of some chromosomes parts. And genes are also annotated on
these alternative sequences. In Uniprot some <em>unreviewed</em>
identifiers can correspond to <em>reviewed</em> proteins.</p>
<p>When available such kind of information is associated to an
<strong>Attribute</strong> node through a <code>has</code> relationship
providing the value of the attribute for the BEID. This information can
also be used to define if a BEID is a <em>preferred</em> identifier for
a BE.</p>
<p>The example below shows the case of the MAPT gene annotated on
different version of human chromosome 17.</p>
<div style="border:solid;overflow:hidden;">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   <span class="st">"MAPT"</span>, from<span class="op">=</span><span class="st">"Gene"</span>, from.source<span class="op">=</span><span class="st">"Symbol"</span>, from.org<span class="op">=</span><span class="st">"human"</span>,</span>
<span>   to.source<span class="op">=</span><span class="st">"Ens_gene"</span>, restricted<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/exploreBe.html">exploreBe</a></span><span class="op">(</span></span>
<span>   <span class="va">mapt</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"to"</span><span class="op">]</span>,</span>
<span>   source<span class="op">=</span><span class="st">"Ens_gene"</span>,</span>
<span>   be<span class="op">=</span><span class="st">"Gene"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"nodes":{"id":[605370,862338,798483,959854,815811,815698,6252668,6182834,6242550,6247609,997475,1016687],"type":["GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","GeneID","BESymbol"],"label":["4137","ENSG00000186868","LRG_660","6893","ENSG00000277956","ENSG00000276155","8152","147077","1651","4138","157140","MAPT"],"database":["EntrezGene","Ens_gene","Ens_gene","HGNC","Ens_gene","Ens_gene","EntrezGene","EntrezGene","EntrezGene","EntrezGene","MIM_GENE",""],"platform":["","","","","","","","","","","",""],"preferred":[true,true,false,false,false,false,false,false,false,false,false,null],"url":["https://www.ncbi.nlm.nih.gov/gene/4137","http://www.ensembl.org/id/ENSG00000186868","http://www.ensembl.org/id/LRG_660","http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=6893","http://www.ensembl.org/id/ENSG00000277956","http://www.ensembl.org/id/ENSG00000276155","https://www.ncbi.nlm.nih.gov/gene/8152","https://www.ncbi.nlm.nih.gov/gene/147077","https://www.ncbi.nlm.nih.gov/gene/1651","https://www.ncbi.nlm.nih.gov/gene/4138","http://www.omim.org/entry/157140",""],"source":["EntrezGene","Ens_gene","Ens_gene","HGNC","Ens_gene","Ens_gene","EntrezGene","EntrezGene","EntrezGene","EntrezGene","MIM_GENE",""],"title":["<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/4137\" target=\"_blank\">4137<\/a><br><emph>EntrezGene<\/emph><br>Preferred<\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/ENSG00000186868\" target=\"_blank\">ENSG00000186868<\/a><br><emph>Ens_gene<\/emph><br>Preferred<\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/LRG_660\" target=\"_blank\">LRG_660<\/a><br><emph>Ens_gene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=6893\" target=\"_blank\">6893<\/a><br><emph>HGNC<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/ENSG00000277956\" target=\"_blank\">ENSG00000277956<\/a><br><emph>Ens_gene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.ensembl.org/id/ENSG00000276155\" target=\"_blank\">ENSG00000276155<\/a><br><emph>Ens_gene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/8152\" target=\"_blank\">8152<\/a><br><emph>EntrezGene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/147077\" target=\"_blank\">147077<\/a><br><emph>EntrezGene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/1651\" target=\"_blank\">1651<\/a><br><emph>EntrezGene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/4138\" target=\"_blank\">4138<\/a><br><emph>EntrezGene<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"http://www.omim.org/entry/157140\" target=\"_blank\">157140<\/a><br><emph>MIM_GENE<\/emph><\/p>","<p><strong>BESymbol<\/strong><br>MAPT<\/p>"],"shape":["dot","dot","dot","dot","dot","dot","dot","dot","dot","dot","dot","box"],"color.background":["#67E781","#E2DBE6","#E2DBE6","#7E58DB","#E2DBE6","#E2DBE6","#67E781","#67E781","#67E781","#67E781","#B694E4","#E4AC77"],"borderWidth":[4,4,1,1,1,1,1,1,1,1,1,1],"borderWidthSelected":[4,4,2,2,2,2,2,2,2,2,2,2],"color.border":["black","black","black","black","black","black","black","black","black","black","black","black"]},"edges":{"id":[13373763,1940269,2224849,1897394,2224850,2224851,2224852,13595317,13602194,13594300,13594667,13432580,2183124,2183114,2183122,2183115,12910227,2359159,2315274,2326564,2326492],"title":["corresponds_to","corresponds_to","corresponds_to","corresponds_to","corresponds_to","corresponds_to","corresponds_to","is_replaced_by","is_replaced_by","is_replaced_by","is_replaced_by","is_associated_to","is_associated_to","is_associated_to","is_associated_to","is_associated_to","is_known_as","is_known_as","is_known_as","is_known_as","is_known_as"],"from":[605370,605370,862338,798483,798483,815811,815698,6252668,6182834,6242550,6247609,997475,997475,997475,997475,997475,605370,862338,798483,815811,815698],"to":[959854,862338,959854,862338,959854,959854,959854,605370,605370,605370,605370,605370,862338,798483,815811,815698,1016687,1016687,1016687,1016687,1016687],"arrows":["to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to"],"dashes":[false,false,false,false,false,false,false,true,true,true,true,true,true,true,true,true,false,false,false,false,false]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot"},"manipulation":{"enabled":false},"interaction":{"selectable":true,"zoomSpeed":1}},"groups":null,"width":null,"height":null,"idselection":{"enabled":true,"style":"width: 150px; height: 26px","useLabels":true,"main":"Select by id","selected":862338},"byselection":{"enabled":false,"style":"width: 150px; height: 26px","multiple":false,"hideColor":"rgba(200,200,200,0.5)","highlight":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);","highlight":{"enabled":true,"hoverNearest":false,"degree":1,"algorithm":"all","hideColor":"rgba(200,200,200,0.5)","labelOnly":true},"collapse":{"enabled":false,"fit":false,"resetHighlight":true,"clusterOptions":null,"keepCoord":true,"labelSuffix":"(cluster)"},"legend":{"width":0.2,"useGroups":false,"position":"left","ncol":2,"stepX":100,"stepY":100,"zoom":true,"nodes":[{"label":"GeneID","shape":"dot","color":null},{"label":"BESymbol","shape":"box","color":"#E4AC77"},{"label":"EntrezGene","shape":"dot","color":"#67E781"},{"label":"Ens_gene","shape":"dot","color":"#E2DBE6"},{"label":"HGNC","shape":"dot","color":"#7E58DB"},{"label":"MIM_GENE","shape":"dot","color":"#B694E4"}],"nodesToDataframe":false}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getBeIds.html">getBeIds</a></span><span class="op">(</span></span>
<span>   be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"Ens_gene"</span>, organism<span class="op">=</span><span class="st">"human"</span>,</span>
<span>   restricted<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>   attributes<span class="op">=</span><span class="fu"><a href="../reference/listDBAttributes.html">listDBAttributes</a></span><span class="op">(</span><span class="st">"Ens_gene"</span><span class="op">)</span>,</span>
<span>   filter<span class="op">=</span><span class="va">mapt</span><span class="op">$</span><span class="va">to</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                id preferred    Gene db.version db.deprecated</span></span>
<span><span class="co">## 1 ENSG00000186868      TRUE 5603384        112         FALSE</span></span>
<span><span class="co">## 2 ENSG00000276155     FALSE 5603384        112         FALSE</span></span>
<span><span class="co">## 3 ENSG00000277956     FALSE 5603384        112         FALSE</span></span>
<span><span class="co">## 4         LRG_660     FALSE 5603384        112         FALSE</span></span>
<span><span class="co">##                       seq_region</span></span>
<span><span class="co">## 1           GRCh38 chromosome 17</span></span>
<span><span class="co">## 2 GRCh38 scaffold HSCHR17_1_CTG5</span></span>
<span><span class="co">## 3 GRCh38 scaffold HSCHR17_2_CTG5</span></span>
<span><span class="co">## 4                    lrg LRG_660</span></span></code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="checking-identifiers">Checking identifiers<a class="anchor" aria-label="anchor" href="#checking-identifiers"></a>
</h3>
<p>The origin of identifiers can be guessed as following.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriId</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"17237"</span>, <span class="st">"105886298"</span>, <span class="st">"76429"</span>, <span class="st">"80985"</span>, <span class="st">"230514"</span>, <span class="st">"66459"</span>,</span>
<span>    <span class="st">"93696"</span>, <span class="st">"72514"</span>, <span class="st">"20352"</span>, <span class="st">"13347"</span>, <span class="st">"100462961"</span>, <span class="st">"100043346"</span>,</span>
<span>    <span class="st">"12400"</span>, <span class="st">"106582"</span>, <span class="st">"19062"</span>, <span class="st">"245607"</span>, <span class="st">"79196"</span>, <span class="st">"16878"</span>, <span class="st">"320727"</span>,</span>
<span>    <span class="st">"230649"</span>, <span class="st">"66880"</span>, <span class="st">"66245"</span>, <span class="st">"103742"</span>, <span class="st">"320145"</span>, <span class="st">"140795"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">idOrigin</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/guessIdScope.html">guessIdScope</a></span><span class="op">(</span><span class="va">oriId</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">idOrigin</span><span class="op">$</span><span class="va">be</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene"</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">idOrigin</span><span class="op">$</span><span class="va">source</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "EntrezGene"</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">idOrigin</span><span class="op">$</span><span class="va">organism</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Mus musculus"</span></span></code></pre>
<p>The best guess is returned as a list but other possible origins are
listed in the <em>details</em> attribute.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">idOrigin</span>, <span class="st">"details"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     be     source     organism nb proportion</span></span>
<span><span class="co">## 1 Gene EntrezGene Mus musculus 25       1.00</span></span>
<span><span class="co">## 2 Gene       HGNC Homo sapiens  3       0.12</span></span>
<span><span class="co">## 3 Gene        MGI Mus musculus  2       0.08</span></span></code></pre>
<p>If the origin of identifiers is already known, it can also be
tested.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/checkBeIds.html">checkBeIds</a></span><span class="op">(</span>ids<span class="op">=</span><span class="va">oriId</span>, be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"mouse"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/checkBeIds.html">checkBeIds</a></span><span class="op">(</span>ids<span class="op">=</span><span class="va">oriId</span>, be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"HGNC"</span>, organism<span class="op">=</span><span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in checkBeIds(ids = oriId, be = "Gene", source = "HGNC", organism =</span></span>
<span><span class="co">## "human"): Could not find 22 IDs among 25!</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="identifier-annotation">Identifier annotation<a class="anchor" aria-label="anchor" href="#identifier-annotation"></a>
</h3>
<p>Identifiers can be annotated with symbols and names according to
available information. The following code returns the most relevant
symbol and the most relevant name for each ID. Source URL can also be
generated with the <code><a href="../reference/getBeIdURL.html">getBeIdURL()</a></code> function.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toShow</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBeIdDescription.html">getBeIdDescription</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">oriId</span>, be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"mouse"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">toShow</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span></span>
<span>        <span class="st">'&lt;a href="%s" target="_blank"&gt;'</span>,</span>
<span>        <span class="fu"><a href="../reference/getBeIdURL.html">getBeIdURL</a></span><span class="op">(</span><span class="va">toShow</span><span class="op">$</span><span class="va">id</span>, <span class="st">"EntrezGene"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="va">toShow</span><span class="op">$</span><span class="va">id</span>,</span>
<span>    <span class="st">'&lt;a&gt;'</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">toShow</span>, escape<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="34%">
<col width="4%">
<col width="46%">
<col width="4%">
<col width="4%">
<col width="5%">
</colgroup>
<thead><tr class="header">
<th align="left">id</th>
<th align="left">symbol</th>
<th align="left">name</th>
<th align="left">preferred</th>
<th align="left">db.version</th>
<th align="left">db.deprecated</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/17237" target="_blank" class="external-link">17237</a><a></a>
</td>
<td align="left">Mgrn1</td>
<td align="left">mahogunin, ring finger 1</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/105886298" target="_blank" class="external-link">105886298</a><a></a>
</td>
<td align="left">Cmc4</td>
<td align="left">C-X9-C motif containing 4</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/76429" target="_blank" class="external-link">76429</a><a></a>
</td>
<td align="left">Lhpp</td>
<td align="left">phospholysine phosphohistidine inorganic pyrophosphate
phosphatase</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/80985" target="_blank" class="external-link">80985</a><a></a>
</td>
<td align="left">Trim44</td>
<td align="left">tripartite motif-containing 44</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/230514" target="_blank" class="external-link">230514</a><a></a>
</td>
<td align="left">Leprot</td>
<td align="left">leptin receptor overlapping transcript</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/66459" target="_blank" class="external-link">66459</a><a></a>
</td>
<td align="left">Pyurf</td>
<td align="left">Pigy upstream reading frame</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/93696" target="_blank" class="external-link">93696</a><a></a>
</td>
<td align="left">Chrac1</td>
<td align="left">chromatin accessibility complex 1</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/72514" target="_blank" class="external-link">72514</a><a></a>
</td>
<td align="left">Fgfbp3</td>
<td align="left">fibroblast growth factor binding protein 3</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/20352" target="_blank" class="external-link">20352</a><a></a>
</td>
<td align="left">Sema4b</td>
<td align="left">sema domain, immunoglobulin domain (Ig), transmembrane
domain (TM) and short cytoplasmic domain, (semaphorin) 4B</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/13347" target="_blank" class="external-link">13347</a><a></a>
</td>
<td align="left">Dffa</td>
<td align="left">DNA fragmentation factor, alpha subunit</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/100462961" target="_blank" class="external-link">100462961</a><a></a>
</td>
<td align="left">Gm16149</td>
<td align="left">predicted gene 16149</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/100043346" target="_blank" class="external-link">100043346</a><a></a>
</td>
<td align="left">Rpl10-ps3</td>
<td align="left">ribosomal protein L10, pseudogene 3</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/12400" target="_blank" class="external-link">12400</a><a></a>
</td>
<td align="left">Cbfb</td>
<td align="left">core binding factor beta</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/106582" target="_blank" class="external-link">106582</a><a></a>
</td>
<td align="left">Nrm</td>
<td align="left">nurim (nuclear envelope membrane protein)</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/19062" target="_blank" class="external-link">19062</a><a></a>
</td>
<td align="left">Inpp5k</td>
<td align="left">inositol polyphosphate 5-phosphatase K</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/245607" target="_blank" class="external-link">245607</a><a></a>
</td>
<td align="left">Gprasp2</td>
<td align="left">G protein-coupled receptor associated sorting protein
2</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/79196" target="_blank" class="external-link">79196</a><a></a>
</td>
<td align="left">Osbpl5</td>
<td align="left">oxysterol binding protein-like 5</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/16878" target="_blank" class="external-link">16878</a><a></a>
</td>
<td align="left">Lif</td>
<td align="left">leukemia inhibitory factor</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/320727" target="_blank" class="external-link">320727</a><a></a>
</td>
<td align="left">Ipo8</td>
<td align="left">importin 8</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/230649" target="_blank" class="external-link">230649</a><a></a>
</td>
<td align="left">Atpaf1</td>
<td align="left">ATP synthase mitochondrial F1 complex assembly factor
1</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/66880" target="_blank" class="external-link">66880</a><a></a>
</td>
<td align="left">Rsrc1</td>
<td align="left">arginine/serine-rich coiled-coil 1</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/66245" target="_blank" class="external-link">66245</a><a></a>
</td>
<td align="left">Hspbp1</td>
<td align="left">HSPA (heat shock 70kDa) binding protein, cytoplasmic
cochaperone 1</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/103742" target="_blank" class="external-link">103742</a><a></a>
</td>
<td align="left">Mien1</td>
<td align="left">migration and invasion enhancer 1</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/320145" target="_blank" class="external-link">320145</a><a></a>
</td>
<td align="left">Sp8</td>
<td align="left">trans-acting transcription factor 8</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">
<a href="https://www.ncbi.nlm.nih.gov/gene/140795" target="_blank" class="external-link">140795</a><a></a>
</td>
<td align="left">P2ry14</td>
<td align="left">purinergic receptor P2Y, G-protein coupled, 14</td>
<td align="left">TRUE</td>
<td align="left">20240516</td>
<td align="left">FALSE</td>
</tr>
</tbody>
</table>
<p>All possible symbols and all possible names for each ID can also be
retrieved using the following functions.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBeIdSymbols.html">getBeIdSymbols</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">oriId</span>, be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          id        symbol canonical direct preferred  entity</span></span>
<span><span class="co">## 1 105886298       Mtcp1nb     FALSE   TRUE      TRUE 2797892</span></span>
<span><span class="co">## 2 105886298       p8MTCP1     FALSE   TRUE      TRUE 2797892</span></span>
<span><span class="co">## 3 105886298          Cmc4      TRUE   TRUE      TRUE 2797892</span></span>
<span><span class="co">## 4     93696          YCL1     FALSE   TRUE      TRUE 2815578</span></span>
<span><span class="co">## 5     93696 2810406L04Rik     FALSE   TRUE      TRUE 2815578</span></span>
<span><span class="co">## 6     93696 2410152E03Rik     FALSE   TRUE      TRUE 2815578</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBeIdNames.html">getBeIdNames</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">oriId</span>, be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          id                                                    name direct</span></span>
<span><span class="co">## 1 105886298                               C-X9-C motif containing 4   TRUE</span></span>
<span><span class="co">## 2     93696                       chromatin accessibility complex 1   TRUE</span></span>
<span><span class="co">## 3     72514              fibroblast growth factor binding protein 3   TRUE</span></span>
<span><span class="co">## 4     12400                                core binding factor beta   TRUE</span></span>
<span><span class="co">## 5     16878                              leukemia inhibitory factor   TRUE</span></span>
<span><span class="co">## 6    245607 G protein-coupled receptor associated sorting protein 2   TRUE</span></span>
<span><span class="co">##   preferred  entity</span></span>
<span><span class="co">## 1      TRUE 2797892</span></span>
<span><span class="co">## 2      TRUE 2815578</span></span>
<span><span class="co">## 3      TRUE 2816491</span></span>
<span><span class="co">## 4      TRUE 2817153</span></span>
<span><span class="co">## 5      TRUE 2828063</span></span>
<span><span class="co">## 6      TRUE 2811916</span></span></code></pre>
<p>Also probes and some biological entities do not have directly
associated symbols or names. These elements can also be annotated
according to information related to relevant genes.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">someProbes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"238834_at"</span>, <span class="st">"1569297_at"</span>, <span class="st">"213021_at"</span>, <span class="st">"225480_at"</span>,</span>
<span>    <span class="st">"216016_at"</span>, <span class="st">"35685_at"</span>, <span class="st">"217969_at"</span>, <span class="st">"211359_s_at"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">toShow</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneDescription.html">getGeneDescription</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">someProbes</span>, be<span class="op">=</span><span class="st">"Probe"</span>, source<span class="op">=</span><span class="st">"GPL570"</span>, organism<span class="op">=</span><span class="st">"human"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">toShow</span>, escape<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="15%">
<col width="14%">
<col width="12%">
<col width="57%">
</colgroup>
<thead><tr class="header">
<th align="left">id</th>
<th align="left">EntrezGene</th>
<th align="left">symbol</th>
<th align="left">name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">238834_at</td>
<td align="left">91807</td>
<td align="left">MYLK3</td>
<td align="left">myosin light chain kinase 3</td>
</tr>
<tr class="even">
<td align="left">1569297_at</td>
<td align="left">731779</td>
<td align="left">LINC01300</td>
<td align="left">long intergenic non-protein coding RNA 1300</td>
</tr>
<tr class="odd">
<td align="left">213021_at</td>
<td align="left">9527</td>
<td align="left">GOSR1</td>
<td align="left">golgi SNAP receptor complex member 1</td>
</tr>
<tr class="even">
<td align="left">225480_at</td>
<td align="left">127687</td>
<td align="left">C1orf122</td>
<td align="left">chromosome 1 open reading frame 122</td>
</tr>
<tr class="odd">
<td align="left">216016_at</td>
<td align="left">114548</td>
<td align="left">NLRP3</td>
<td align="left">NLR family pyrin domain containing 3</td>
</tr>
<tr class="even">
<td align="left">35685_at</td>
<td align="left">6015</td>
<td align="left">RING1</td>
<td align="left">ring finger protein 1</td>
</tr>
<tr class="odd">
<td align="left">217969_at</td>
<td align="left">738</td>
<td align="left">VPS51</td>
<td align="left">VPS51 subunit of GARP complex</td>
</tr>
<tr class="even">
<td align="left">211359_s_at</td>
<td align="left">4988</td>
<td align="left">OPRM1</td>
<td align="left">opioid receptor mu 1</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="products-of-molecular-biology-processes">Products of molecular biology processes<a class="anchor" aria-label="anchor" href="#products-of-molecular-biology-processes"></a>
</h3>
<p>The BED data model has beeing built to fulfill molecular biology
processes:</p>
<ul>
<li>
<strong>is_expressed_as</strong> relationships correspond to the
transcription process.</li>
<li>
<strong>is_translated_in</strong> relationships correspond to the
translation process.</li>
<li>
<strong>codes_for</strong> is a fuzzy relationship allowing the
mapping of genes on object not necessary corresonpding to the same kind
of biological molecule.</li>
</ul>
<p>These processes are described in different databases with different
level of granularity. For exemple, Ensembl provides possible transcripts
for each gene specifying which one of them is canonical.</p>
<p>The following functions are used to retrieve direct products or
direct origins of molecular biology processes.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getDirectProduct.html">getDirectProduct</a></span><span class="op">(</span><span class="st">"ENSG00000145335"</span>, process<span class="op">=</span><span class="st">"is_expressed_as"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             origin  osource         product        psource canonical</span></span>
<span><span class="co">## 1  ENSG00000145335 Ens_gene ENST00000505199 Ens_transcript     FALSE</span></span>
<span><span class="co">## 2  ENSG00000145335 Ens_gene ENST00000673766 Ens_transcript     FALSE</span></span>
<span><span class="co">## 3  ENSG00000145335 Ens_gene ENST00000674129 Ens_transcript     FALSE</span></span>
<span><span class="co">## 4  ENSG00000145335 Ens_gene ENST00000508895 Ens_transcript     FALSE</span></span>
<span><span class="co">## 5  ENSG00000145335 Ens_gene ENST00000673718 Ens_transcript     FALSE</span></span>
<span><span class="co">## 6  ENSG00000145335 Ens_gene ENST00000611107 Ens_transcript     FALSE</span></span>
<span><span class="co">## 7  ENSG00000145335 Ens_gene ENST00000345009 Ens_transcript     FALSE</span></span>
<span><span class="co">## 8  ENSG00000145335 Ens_gene ENST00000394986 Ens_transcript     FALSE</span></span>
<span><span class="co">## 9  ENSG00000145335 Ens_gene ENST00000618500 Ens_transcript     FALSE</span></span>
<span><span class="co">## 10 ENSG00000145335 Ens_gene ENST00000420646 Ens_transcript     FALSE</span></span>
<span><span class="co">## 11 ENSG00000145335 Ens_gene ENST00000394989 Ens_transcript     FALSE</span></span>
<span><span class="co">## 12 ENSG00000145335 Ens_gene ENST00000336904 Ens_transcript     FALSE</span></span>
<span><span class="co">## 13 ENSG00000145335 Ens_gene ENST00000506244 Ens_transcript     FALSE</span></span>
<span><span class="co">## 14 ENSG00000145335 Ens_gene ENST00000502987 Ens_transcript     FALSE</span></span>
<span><span class="co">## 15 ENSG00000145335 Ens_gene ENST00000506691 Ens_transcript     FALSE</span></span>
<span><span class="co">## 16 ENSG00000145335 Ens_gene ENST00000673902 Ens_transcript     FALSE</span></span>
<span><span class="co">## 17 ENSG00000145335 Ens_gene ENST00000394991 Ens_transcript      TRUE</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getDirectProduct.html">getDirectProduct</a></span><span class="op">(</span><span class="st">"ENST00000336904"</span>, process<span class="op">=</span><span class="st">"is_translated_in"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            origin        osource         product         psource canonical</span></span>
<span><span class="co">## 1 ENST00000336904 Ens_transcript ENSP00000338345 Ens_translation      TRUE</span></span></code></pre>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getDirectOrigin.html">getDirectOrigin</a></span><span class="op">(</span><span class="st">"NM_001146055"</span>, process<span class="op">=</span><span class="st">"is_expressed_as"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   origin    osource      product psource canonical</span></span>
<span><span class="co">## 1   6622 EntrezGene NM_001146055  RefSeq     FALSE</span></span></code></pre>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
</div>
<div class="section level2">
<h2 id="converting-identifiers">Converting identifiers<a class="anchor" aria-label="anchor" href="#converting-identifiers"></a>
</h2>
<div class="section level3">
<h3 id="same-entity-and-same-organism-from-one-source-to-another">Same entity and same organism: from one source to another<a class="anchor" aria-label="anchor" href="#same-entity-and-same-organism-from-one-source-to-another"></a>
</h3>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">oriId</span>,</span>
<span>    from<span class="op">=</span><span class="st">"Gene"</span>,</span>
<span>    from.source<span class="op">=</span><span class="st">"EntrezGene"</span>,</span>
<span>    from.org<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    to.source<span class="op">=</span><span class="st">"Ens_gene"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    prefFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        from                 to to.preferred to.entity</span></span>
<span><span class="co">## 1     19062 ENSMUSG00000006127         TRUE   2790077</span></span>
<span><span class="co">## 2     13347 ENSMUSG00000028974         TRUE   2791260</span></span>
<span><span class="co">## 3    103742 ENSMUSG00000002580         TRUE   2792328</span></span>
<span><span class="co">## 4    320145 ENSMUSG00000048562         TRUE   2795236</span></span>
<span><span class="co">## 5 105886298 ENSMUSG00000090110         TRUE   2797892</span></span>
<span><span class="co">## 6    320727 ENSMUSG00000040029         TRUE   2799905</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="same-organism-from-one-entity-to-another">Same organism: from one entity to another<a class="anchor" aria-label="anchor" href="#same-organism-from-one-entity-to-another"></a>
</h3>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">oriId</span>,</span>
<span>    from<span class="op">=</span><span class="st">"Gene"</span>,</span>
<span>    from.source<span class="op">=</span><span class="st">"EntrezGene"</span>,</span>
<span>    from.org<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    to<span class="op">=</span><span class="st">"Peptide"</span>,</span>
<span>    to.source<span class="op">=</span><span class="st">"Ens_translation"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    prefFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         from                 to to.preferred to.entity</span></span>
<span><span class="co">## 1      19062 ENSMUSP00000119996         TRUE   3761978</span></span>
<span><span class="co">## 2      19062 ENSMUSP00000121060         TRUE   3761986</span></span>
<span><span class="co">## 3     320145 ENSMUSP00000152523         TRUE   3767751</span></span>
<span><span class="co">## 4  105886298 ENSMUSP00000033543         TRUE   3770795</span></span>
<span><span class="co">## 10 105886298 ENSMUSP00000112753         TRUE   3770795</span></span>
<span><span class="co">## 7  105886298 ENSMUSP00000117381         TRUE   3770795</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="from-one-organism-to-another">From one organism to another<a class="anchor" aria-label="anchor" href="#from-one-organism-to-another"></a>
</h3>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>    ids<span class="op">=</span><span class="va">oriId</span>,</span>
<span>    from<span class="op">=</span><span class="st">"Gene"</span>,</span>
<span>    from.source<span class="op">=</span><span class="st">"EntrezGene"</span>,</span>
<span>    from.org<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    to<span class="op">=</span><span class="st">"Peptide"</span>,</span>
<span>    to.source<span class="op">=</span><span class="st">"Ens_translation"</span>,</span>
<span>    to.org<span class="op">=</span><span class="st">"human"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    prefFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      from              to to.preferred to.entity</span></span>
<span><span class="co">## 34  17237 ENSP00000468171         TRUE   2385263</span></span>
<span><span class="co">## 35  17237 ENSP00000443810         TRUE   2385494</span></span>
<span><span class="co">## 36  17237 ENSP00000467414         TRUE   2385496</span></span>
<span><span class="co">## 37  17237 ENSP00000464811         TRUE   2385515</span></span>
<span><span class="co">## 38  17237 ENSP00000465249         TRUE   2385516</span></span>
<span><span class="co">## 19 103742 ENSP00000377778         TRUE   2393233</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="converting-lists-of-identifiers">Converting lists of identifiers<a class="anchor" aria-label="anchor" href="#converting-lists-of-identifiers"></a>
</h3>
<p>List of identifiers can be converted the following way. Only
converted IDs are returned in this case.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">humanEnsPeptides</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convBeIdLists.html">convBeIdLists</a></span><span class="op">(</span></span>
<span>    idList<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">oriId</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, b<span class="op">=</span><span class="va">oriId</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    from<span class="op">=</span><span class="st">"Gene"</span>,</span>
<span>    from.source<span class="op">=</span><span class="st">"EntrezGene"</span>,</span>
<span>    from.org<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    to<span class="op">=</span><span class="st">"Peptide"</span>,</span>
<span>    to.source<span class="op">=</span><span class="st">"Ens_translation"</span>,</span>
<span>    to.org<span class="op">=</span><span class="st">"human"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    prefFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">humanEnsPeptides</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   a   b </span></span>
<span><span class="co">##  21 117</span></span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">humanEnsPeptides</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $a</span></span>
<span><span class="co">## [1] "ENSP00000468171" "ENSP00000443810" "ENSP00000467414" "ENSP00000464811"</span></span>
<span><span class="co">## [5] "ENSP00000465249" "ENSP00000497944"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $b</span></span>
<span><span class="co">## [1] "ENSP00000377778" "ENSP00000462998" "ENSP00000466569" "ENSP00000466749"</span></span>
<span><span class="co">## [5] "ENSP00000465876" "ENSP00000465690"</span></span></code></pre>
<div class="section level4">
<h4 id="beidlist">BEIDList<a class="anchor" aria-label="anchor" href="#beidlist"></a>
</h4>
<p><code>BEIDList</code> objects are used to manage lists of BEID with
an attached explicit scope, and metadata provided in a data frame. The
<code><a href="../reference/focusOnScope.html">focusOnScope()</a></code> function is used to easily convert such
object to another scope. For example, in the code below, Entrez gene
identifiers are converted in Ensembl identifiers.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">entrezGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BEIDList.html">BEIDList</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">oriId</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, b<span class="op">=</span><span class="va">oriId</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>   scope<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"EntrezGene"</span>, organism<span class="op">=</span><span class="st">"Mus musculus"</span><span class="op">)</span>,</span>
<span>   metadata<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>      .lname<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>,</span>
<span>      description<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Identifiers in a"</span>, <span class="st">"Identifiers in b"</span><span class="op">)</span>,</span>
<span>      stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span></span>
<span>   <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">entrezGenes</span></span></code></pre></div>
<pre><code><span><span class="co">## BEIDList of 2 elements gathering 25 BEIDs in total</span></span>
<span><span class="co">##    - Scope: be="Gene", source="EntrezGene", organism="Mus musculus"</span></span>
<span><span class="co">##    - Metadata fields: ".lname", "description"</span></span></code></pre>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">entrezGenes</span><span class="op">$</span><span class="va">a</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "17237"     "105886298" "76429"     "80985"     "230514"</span></span></code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ensemblGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/focusOnScope.html">focusOnScope</a></span><span class="op">(</span><span class="va">entrezGenes</span>, source<span class="op">=</span><span class="st">"Ens_gene"</span><span class="op">)</span></span>
<span><span class="va">ensemblGenes</span><span class="op">$</span><span class="va">a</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ENSMUSG00000090110" "ENSMUSG00000022517" "ENSMUSG00000035212"</span></span>
<span><span class="co">## [4] "ENSMUSG00000030946" "ENSMUSG00000027189"</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="converting-data-frames">Converting data frames<a class="anchor" aria-label="anchor" href="#converting-data-frames"></a>
</h3>
<p>IDs in data frames can also be converted.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toConv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">25</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">toConv</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">oriId</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convDfBeIds.html">convDfBeIds</a></span><span class="op">(</span></span>
<span>    df<span class="op">=</span><span class="va">toConv</span>,</span>
<span>    from<span class="op">=</span><span class="st">"Gene"</span>,</span>
<span>    from.source<span class="op">=</span><span class="st">"EntrezGene"</span>,</span>
<span>    from.org<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>    to.source<span class="op">=</span><span class="st">"Ens_gene"</span>,</span>
<span>    restricted<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    prefFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   a           b conv.from            conv.to</span></span>
<span><span class="co">## 1 1 0.080750138     17237 ENSMUSG00000022517</span></span>
<span><span class="co">## 2 2 0.834333037 105886298 ENSMUSG00000090110</span></span>
<span><span class="co">## 3 3 0.600760886     76429 ENSMUSG00000030946</span></span>
<span><span class="co">## 4 4 0.157208442     80985 ENSMUSG00000027189</span></span>
<span><span class="co">## 5 5 0.007399441    230514 ENSMUSG00000035212</span></span>
<span><span class="co">## 6 6 0.466393497     66459 ENSMUSG00000043162</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="explore-convertion-shortest-path-between-two-identifiers">Explore convertion shortest path between two identifiers<a class="anchor" aria-label="anchor" href="#explore-convertion-shortest-path-between-two-identifiers"></a>
</h3>
<p>Because the conversion process takes into account several resources,
it might be useful to explore the path between two identifiers which
have been mapped. This can be achieved by the
<code>exploreConvPath</code> function.</p>
<div style="border:solid;overflow:hidden;">
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">from.id</span> <span class="op">&lt;-</span> <span class="st">"ILMN_1220595"</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   ids<span class="op">=</span><span class="va">from.id</span>, from<span class="op">=</span><span class="st">"Probe"</span>, from.source<span class="op">=</span><span class="st">"GPL6885"</span>, from.org<span class="op">=</span><span class="st">"mouse"</span>,</span>
<span>   to<span class="op">=</span><span class="st">"Peptide"</span>, to.source<span class="op">=</span><span class="st">"Uniprot"</span>, to.org<span class="op">=</span><span class="st">"human"</span>,</span>
<span>   prefFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">##           from     to to.preferred to.entity</span></span>
<span><span class="co">## 1 ILMN_1220595 Q16552         TRUE   8712579</span></span></code></pre>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/exploreConvPath.html">exploreConvPath</a></span><span class="op">(</span></span>
<span>   from.id<span class="op">=</span><span class="va">from.id</span>, from<span class="op">=</span><span class="st">"Probe"</span>, from.source<span class="op">=</span><span class="st">"GPL6885"</span>,</span>
<span>   to.id<span class="op">=</span><span class="va">res</span><span class="op">$</span><span class="va">to</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, to<span class="op">=</span><span class="st">"Peptide"</span>, to.source<span class="op">=</span><span class="st">"Uniprot"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"nodes":{"id":[13151837,2902946,604594,1794891,2466714,3436143,2507095],"type":["ProbeID","GeneID","GeneID","TranscriptID","PeptideID","TranscriptID","PeptideID"],"label":["ILMN_1220595","16171","3605","NM_002190","Q16552","NM_010552","NP_002181"],"database":["","EntrezGene","EntrezGene","RefSeq","Uniprot","RefSeq","RefSeq_peptide"],"preferred":[null,true,true,true,true,true,true],"platform":["GPL6885","","","","","",""],"url":[null,"https://www.ncbi.nlm.nih.gov/gene/16171","https://www.ncbi.nlm.nih.gov/gene/3605","https://www.ncbi.nlm.nih.gov/nuccore/NM_002190","http://www.uniprot.org/uniprot/Q16552","https://www.ncbi.nlm.nih.gov/nuccore/NM_010552","https://www.ncbi.nlm.nih.gov/protein/NP_002181"],"symbol":["","Il17a","IL17A","","IL17_HUMAN","",""],"source":["GPL6885","EntrezGene","EntrezGene","RefSeq","Uniprot","RefSeq","RefSeq_peptide"],"title":["<p><strong>ProbeID<\/strong><br>ILMN_1220595<br><emph>GPL6885<\/emph><\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/16171\" target=\"_blank\">16171<\/a><br><emph>EntrezGene<\/emph><br>Preferred<br>Il17a<\/p>","<p><strong>GeneID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/gene/3605\" target=\"_blank\">3605<\/a><br><emph>EntrezGene<\/emph><br>Preferred<br>IL17A<\/p>","<p><strong>TranscriptID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/nuccore/NM_002190\" target=\"_blank\">NM_002190<\/a><br><emph>RefSeq<\/emph><br>Preferred<\/p>","<p><strong>PeptideID<\/strong><br><a href=\"http://www.uniprot.org/uniprot/Q16552\" target=\"_blank\">Q16552<\/a><br><emph>Uniprot<\/emph><br>Preferred<br>IL17_HUMAN<\/p>","<p><strong>TranscriptID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/nuccore/NM_010552\" target=\"_blank\">NM_010552<\/a><br><emph>RefSeq<\/emph><br>Preferred<\/p>","<p><strong>PeptideID<\/strong><br><a href=\"https://www.ncbi.nlm.nih.gov/protein/NP_002181\" target=\"_blank\">NP_002181<\/a><br><emph>RefSeq_peptide<\/emph><br>Preferred<\/p>"],"shape":["triangle","dot","dot","diamond","square","diamond","square"],"color.background":["#E2B845","#E4AC77","#E4AC77","#67E781","#E2DBE6","#67E781","#7E58DB"],"borderWidth":[1,4,4,4,4,4,4],"borderWidthSelected":[2,4,4,4,4,4,4],"color.border":["black","black","black","black","black","black","black"]},"edges":{"id":[30161494,17409236,29319315,14540062,15687663,20599007],"title":["targets","is_expressed_as","is_homolog_of","is_expressed_as","is_translated_in","corresponds_to"],"from":[13151837,2902946,2902946,604594,1794891,2466714],"to":[3436143,3436143,604594,1794891,2507095,2507095],"arrows":["","","","","",""],"dashes":[false,false,false,false,false,false],"color":["black","black","red","black","black","black"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot"},"manipulation":{"enabled":false},"interaction":{"selectable":true,"zoomSpeed":1}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false,"style":"width: 150px; height: 26px","useLabels":true,"main":"Select by id"},"byselection":{"enabled":false,"style":"width: 150px; height: 26px","multiple":false,"hideColor":"rgba(200,200,200,0.5)","highlight":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);","highlight":{"enabled":true,"hoverNearest":false,"degree":1,"algorithm":"all","hideColor":"rgba(200,200,200,0.5)","labelOnly":true},"collapse":{"enabled":false,"fit":false,"resetHighlight":true,"clusterOptions":null,"keepCoord":true,"labelSuffix":"(cluster)"},"legend":{"width":0.3,"useGroups":false,"position":"left","ncol":3,"stepX":100,"stepY":100,"zoom":true,"nodes":[{"label":"GeneID","shape":"dot","color":null},{"label":"TranscriptID","shape":"diamond","color":null},{"label":"PeptideID","shape":"square","color":null},{"label":"ProbeID","shape":"triangle","color":null},{"label":"GPL6885","shape":"dot","color":"#E2B845"},{"label":"EntrezGene","shape":"dot","color":"#E4AC77"},{"label":"RefSeq","shape":"dot","color":"#67E781"},{"label":"Uniprot","shape":"dot","color":"#E2DBE6"},{"label":"RefSeq_peptide","shape":"dot","color":"#7E58DB"}],"nodesToDataframe":false}},"evals":[],"jsHooks":[]}</script>
</div>
<p>The figure above shows how the ILMN_1220595 ProbeID, targeting the
mouse NM_010552 transcript, can be associated to the Q16552 human
protein ID in Uniprot.</p>
</div>
<div class="section level3">
<h3 id="notes-about-converting-from-and-to-gene-symbols">Notes about converting from and to gene symbols<a class="anchor" aria-label="anchor" href="#notes-about-converting-from-and-to-gene-symbols"></a>
</h3>
<p>Canonical and non-canonical symbols are associated to genes. In some
cases the same symbol (canonical or not) can be associated to several
genes. This can lead to ambiguous mapping. The strategy to apply for
such mapping depends on the aim of the user and his knowledge about the
origin of the symbols to consider.</p>
<p>The complete mapping between Ensembl gene identifiers and symbols is
retrieved by using the <code>getBeIDSymbolTable</code> function.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compMap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBeIdSymbolTable.html">getBeIdSymbolTable</a></span><span class="op">(</span></span>
<span>   be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"Ens_gene"</span>, organism<span class="op">=</span><span class="st">"rat"</span>,</span>
<span>   restricted<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">compMap</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 139938      6</span></span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">compMap</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   id       symbol canonical direct preferred  entity</span></span>
<span><span class="co">## 1 ENSRNOG00000016565    LOC365179     FALSE   TRUE      TRUE 4000180</span></span>
<span><span class="co">## 2 ENSRNOG00000016565   RGD1565485     FALSE   TRUE      TRUE 4000180</span></span>
<span><span class="co">## 3 ENSRNOG00000016565       Znf524     FALSE   TRUE      TRUE 4000180</span></span>
<span><span class="co">## 4 ENSRNOG00000016565       Zfp524      TRUE   TRUE      TRUE 4000180</span></span>
<span><span class="co">## 5 ENSRNOG00000017878       Aldhpb     FALSE   TRUE      TRUE 7830536</span></span>
<span><span class="co">## 6 ENSRNOG00000017878 LOC108349824     FALSE   TRUE      TRUE 7830536</span></span></code></pre>
<p>The canonical field indicates if the symbol is canonical for the
identifier. The direct field indicates if the symbol is directly
associated to the identifier or indirectly through a relationship with
another identifier.</p>
<p>As an example, let’s consider the “Snca” symbol in rat. As shown
below, this symbol is associated to 2 genes; it is canonical for one
gene and not for another. These 2 genes are also associated to other
symbols.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sncaEid</span> <span class="op">&lt;-</span> <span class="va">compMap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">compMap</span><span class="op">$</span><span class="va">symbol</span><span class="op">==</span><span class="st">"Snca"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">sncaEid</span></span></code></pre></div>
<pre><code><span><span class="co">##                       id symbol canonical direct preferred  entity</span></span>
<span><span class="co">## 18141 ENSRNOG00000029408   Snca     FALSE   TRUE      TRUE 4021424</span></span>
<span><span class="co">## 67789 ENSRNOG00000008656   Snca      TRUE   TRUE      TRUE 4002847</span></span></code></pre>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compMap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">compMap</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sncaEid</span><span class="op">$</span><span class="va">id</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                       id    symbol canonical direct preferred  entity</span></span>
<span><span class="co">## 18140 ENSRNOG00000029408 LOC317274     FALSE   TRUE      TRUE 4021424</span></span>
<span><span class="co">## 18141 ENSRNOG00000029408      Snca     FALSE   TRUE      TRUE 4021424</span></span>
<span><span class="co">## 18142 ENSRNOG00000029408   Mageb16      TRUE   TRUE      TRUE 4021424</span></span>
<span><span class="co">## 67788 ENSRNOG00000008656 MGC105443     FALSE   TRUE      TRUE 4002847</span></span>
<span><span class="co">## 67789 ENSRNOG00000008656      Snca      TRUE   TRUE      TRUE 4002847</span></span></code></pre>
<p>The <code>getBeIdDescription</code> function described before,
reports only one symbol for each identifier. Canonical and direct
symbols are prioritized.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getBeIdDescription.html">getBeIdDescription</a></span><span class="op">(</span></span>
<span>   <span class="va">sncaEid</span><span class="op">$</span><span class="va">id</span>,</span>
<span>   be<span class="op">=</span><span class="st">"Gene"</span>, source<span class="op">=</span><span class="st">"Ens_gene"</span>, organism<span class="op">=</span><span class="st">"rat"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                    id  symbol                   name preferred</span></span>
<span><span class="co">## ENSRNOG00000029408 ENSRNOG00000029408 Mageb16 MAGE family member B16      TRUE</span></span>
<span><span class="co">## ENSRNOG00000008656 ENSRNOG00000008656    Snca        synuclein alpha      TRUE</span></span>
<span><span class="co">##                    db.version db.deprecated</span></span>
<span><span class="co">## ENSRNOG00000029408        112         FALSE</span></span>
<span><span class="co">## ENSRNOG00000008656        112         FALSE</span></span></code></pre>
<p>The <code>convBeIds</code> works differently in order to provide a
mapping as exhaustive as possible. If a symbol is associated to several
input identifiers, non-canonical associations with this symbol are
removed if a canonical association exists for any other identifier. This
can lead to inconsistent results, depending on the user input, as show
below.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   <span class="va">sncaEid</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>   from<span class="op">=</span><span class="st">"Gene"</span>, from.source<span class="op">=</span><span class="st">"Ens_gene"</span>, from.org<span class="op">=</span><span class="st">"rat"</span>,</span>
<span>   to.source<span class="op">=</span><span class="st">"Symbol"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 from        to to.preferred to.entity</span></span>
<span><span class="co">## 3 ENSRNOG00000029408 LOC317274           NA   4021424</span></span>
<span><span class="co">## 2 ENSRNOG00000029408   Mageb16           NA   4021424</span></span>
<span><span class="co">## 1 ENSRNOG00000029408      Snca           NA   4021424</span></span></code></pre>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   <span class="va">sncaEid</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>   from<span class="op">=</span><span class="st">"Gene"</span>, from.source<span class="op">=</span><span class="st">"Ens_gene"</span>, from.org<span class="op">=</span><span class="st">"rat"</span>,</span>
<span>   to.source<span class="op">=</span><span class="st">"Symbol"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 from        to to.preferred to.entity</span></span>
<span><span class="co">## 2 ENSRNOG00000008656 MGC105443           NA   4002847</span></span>
<span><span class="co">## 1 ENSRNOG00000008656      Snca           NA   4002847</span></span></code></pre>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   <span class="va">sncaEid</span><span class="op">$</span><span class="va">id</span>,</span>
<span>   from<span class="op">=</span><span class="st">"Gene"</span>, from.source<span class="op">=</span><span class="st">"Ens_gene"</span>, from.org<span class="op">=</span><span class="st">"rat"</span>,</span>
<span>   to.source<span class="op">=</span><span class="st">"Symbol"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 from        to to.preferred to.entity</span></span>
<span><span class="co">## 2 ENSRNOG00000008656 MGC105443           NA   4002847</span></span>
<span><span class="co">## 1 ENSRNOG00000008656      Snca           NA   4002847</span></span>
<span><span class="co">## 5 ENSRNOG00000029408 LOC317274           NA   4021424</span></span>
<span><span class="co">## 4 ENSRNOG00000029408   Mageb16           NA   4021424</span></span></code></pre>
<p>In the example above, when the query is run for each identifier
independently, the association to the “Snca” symbol is reported for
both. However, when running the same query with the 2 identifiers at the
same time, the “Snca” symbol is reported only for one gene corresponding
to the canonical association. An additional filter can be used to only
keep canonical symbols:</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   <span class="va">sncaEid</span><span class="op">$</span><span class="va">id</span>,</span>
<span>   from<span class="op">=</span><span class="st">"Gene"</span>, from.source<span class="op">=</span><span class="st">"Ens_gene"</span>, from.org<span class="op">=</span><span class="st">"rat"</span>,</span>
<span>   to.source<span class="op">=</span><span class="st">"Symbol"</span>,</span>
<span>   canonical<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 from      to to.preferred to.entity</span></span>
<span><span class="co">## 1 ENSRNOG00000008656    Snca           NA   4002847</span></span>
<span><span class="co">## 2 ENSRNOG00000029408 Mageb16           NA   4021424</span></span></code></pre>
<p>Finally, as shown below, when running the query the other way, “Snca”
is only associated to the gene for which it is the canonical symbol.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/convBeIds.html">convBeIds</a></span><span class="op">(</span></span>
<span>   <span class="st">"Snca"</span>,</span>
<span>   from<span class="op">=</span><span class="st">"Gene"</span>, from.source<span class="op">=</span><span class="st">"Symbol"</span>, from.org<span class="op">=</span><span class="st">"rat"</span>,</span>
<span>   to.source<span class="op">=</span><span class="st">"Ens_gene"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   from                 to to.preferred to.entity</span></span>
<span><span class="co">## 1 Snca ENSRNOG00000008656         TRUE   4002847</span></span></code></pre>
<p><strong> Therefore, the user should chose the function to use with
care when needing to convert from or to gene symbol. </strong></p>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
</div>
<div class="section level2">
<h2 id="an-interactive-dictionary-shiny-module">An interactive dictionary: Shiny module<a class="anchor" aria-label="anchor" href="#an-interactive-dictionary-shiny-module"></a>
</h2>
<p>IDs, symbols and names can be seeked without knowing the original
biological entity or probe. Then the results can be converted to the
context of interest.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">searched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/searchBeid.html">searchBeid</a></span><span class="op">(</span><span class="st">"sv2A"</span><span class="op">)</span></span>
<span><span class="va">toTake</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">searched</span><span class="op">$</span><span class="va">organism</span><span class="op">==</span><span class="st">"Homo sapiens"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">relIds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneIDsToAllScopes.html">geneIDsToAllScopes</a></span><span class="op">(</span></span>
<span>  geneids<span class="op">=</span><span class="va">searched</span><span class="op">$</span><span class="va">GeneID</span><span class="op">[</span><span class="va">toTake</span><span class="op">]</span>,</span>
<span>  source<span class="op">=</span><span class="va">searched</span><span class="op">$</span><span class="va">Gene_source</span><span class="op">[</span><span class="va">toTake</span><span class="op">]</span>,</span>
<span>  organism<span class="op">=</span><span class="va">searched</span><span class="op">$</span><span class="va">organism</span><span class="op">[</span><span class="va">toTake</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>A Shiny gadget integrating these two function has been developped and
is also available as an Rstudio addins.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">relIds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findBeids.html">findBeids</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>It relies on a Shiny module (<code><a href="../reference/beidsServer.html">beidsServer()</a></code> and
<code><a href="../reference/beidsServer.html">beidsUI()</a></code> functions) made to facilitate the development of
applications focused on biological entity related information. The code
below shows a minimum example of such an application.</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://shiny.posit.co/" class="external-link">shiny</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patzaw.github.io/BED/">BED</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">ui</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html" class="external-link">fluidPage</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="../reference/beidsServer.html">beidsUI</a></span><span class="op">(</span><span class="st">"be"</span><span class="op">)</span>,</span>
<span>   <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html" class="external-link">fluidRow</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/column.html" class="external-link">column</a></span><span class="op">(</span></span>
<span>         <span class="fl">12</span>,</span>
<span>         <span class="va">tags</span><span class="op">$</span><span class="fu">br</span><span class="op">(</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rstudio.github.io/htmltools/reference/builder.html" class="external-link">h3</a></span><span class="op">(</span><span class="st">"Selected gene entities"</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/pkg/DT/man/dataTableOutput.html" class="external-link">DTOutput</a></span><span class="op">(</span><span class="st">"result"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>   <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">server</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">input</span>, <span class="va">output</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">found</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/beidsServer.html">beidsServer</a></span><span class="op">(</span><span class="st">"be"</span>, toGene<span class="op">=</span><span class="cn">TRUE</span>, multiple<span class="op">=</span><span class="cn">TRUE</span>, tableHeight<span class="op">=</span><span class="fl">250</span><span class="op">)</span></span>
<span>    <span class="va">output</span><span class="op">$</span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DT/man/dataTableOutput.html" class="external-link">renderDT</a></span><span class="op">(</span><span class="op">{</span></span>
<span>       <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/req.html" class="external-link">req</a></span><span class="op">(</span><span class="fu">found</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>       <span class="va">toRet</span> <span class="op">&lt;-</span> <span class="fu">found</span><span class="op">(</span><span class="op">)</span></span>
<span>       <span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">toRet</span>, rownames<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/shinyApp.html" class="external-link">shinyApp</a></span><span class="op">(</span>ui <span class="op">=</span> <span class="va">ui</span>, server <span class="op">=</span> <span class="va">server</span><span class="op">)</span></span></code></pre></div>
<!----------------------------------------------------------------------------->
<!----------------------------------------------------------------------------->
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Red Hat Enterprise Linux</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so;  LAPACK version 3.8.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Brussels</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BED_1.5.2        visNetwork_2.1.2 neo2R_2.4.2      knitr_1.46      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.9        utf8_1.2.4        generics_0.1.3    stringi_1.8.4    </span></span>
<span><span class="co">##  [5] digest_0.6.35     magrittr_2.0.3    evaluate_0.23     fastmap_1.2.0    </span></span>
<span><span class="co">##  [9] jsonlite_1.8.8    promises_1.3.0    httr_1.4.7        purrr_1.0.2      </span></span>
<span><span class="co">## [13] fansi_1.0.6       textshaping_0.3.7 jquerylib_0.1.4   cli_3.6.2        </span></span>
<span><span class="co">## [17] shiny_1.8.1.1     rlang_1.1.3       withr_3.0.0       base64enc_0.1-3  </span></span>
<span><span class="co">## [21] cachem_1.1.0      yaml_2.3.8        tools_4.3.0       memoise_2.0.1    </span></span>
<span><span class="co">## [25] dplyr_1.1.4       httpuv_1.6.15     DT_0.33           curl_5.2.1       </span></span>
<span><span class="co">## [29] vctrs_0.6.5       R6_2.5.1          mime_0.12         lifecycle_1.0.4  </span></span>
<span><span class="co">## [33] stringr_1.5.1     fs_1.6.4          htmlwidgets_1.6.4 miniUI_0.1.1.1   </span></span>
<span><span class="co">## [37] ragg_1.3.2        pkgconfig_2.0.3   desc_1.4.3        pkgdown_2.0.9    </span></span>
<span><span class="co">## [41] pillar_1.9.0      bslib_0.7.0       later_1.3.2       glue_1.7.0       </span></span>
<span><span class="co">## [45] Rcpp_1.0.12       systemfonts_1.1.0 xfun_0.44         tibble_3.2.1     </span></span>
<span><span class="co">## [49] tidyselect_1.2.1  rstudioapi_0.16.0 xtable_1.8-4      htmltools_0.5.8.1</span></span>
<span><span class="co">## [53] rmarkdown_2.26    compiler_4.3.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://patzaw.github.io" class="external-link">Patrice Godard</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
