<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ BED</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BED</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.13</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/BED.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/patzaw/BED/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="BED.html">BED</a></code> 
        </dt>
        <dd>Biological Entity Dictionary (BED)</dd>
      </dl><dl><dt>
          
          <code><a href="BEIDList.html">BEIDList()</a></code> 
        </dt>
        <dd>Create a BEIDList</dd>
      </dl><dl><dt>
          
          <code><a href="BEIDs.html">BEIDs()</a></code> 
        </dt>
        <dd>Get the BEIDs from an object</dd>
      </dl><dl><dt>
          
          <code><a href="beIDsToAllScopes.html">beIDsToAllScopes()</a></code> 
        </dt>
        <dd>Find all BEID and ProbeID corresponding to a BE</dd>
      </dl><dl><dt>
          
          <code><a href="bedCall.html">bedCall()</a></code> 
        </dt>
        <dd>Call a function on the BED graph</dd>
      </dl><dl><dt>
          
          <code><a href="bedImport.html">bedImport()</a></code> 
        </dt>
        <dd>Feeding BED: Imports a data.frame in the BED graph database</dd>
      </dl><dl><dt>
          
          <code><a href="beidsServer.html">beidsServer()</a></code> <code><a href="beidsServer.html">beidsUI()</a></code> 
        </dt>
        <dd>Shiny module for searching BEIDs</dd>
      </dl><dl><dt>
          
          <code><a href="cacheBedCall.html">cacheBedCall()</a></code> 
        </dt>
        <dd>Cached neo4j call</dd>
      </dl><dl><dt>
          
          <code><a href="cacheBedResult.html">cacheBedResult()</a></code> 
        </dt>
        <dd>Put a BED query result in cache</dd>
      </dl><dl><dt>
          
          <code><a href="checkBeIds.html">checkBeIds()</a></code> 
        </dt>
        <dd>Check biological entities (BE) identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="checkBedCache.html">checkBedCache()</a></code> 
        </dt>
        <dd>Check BED cache</dd>
      </dl><dl><dt>
          
          <code><a href="checkBedConn.html">checkBedConn()</a></code> 
        </dt>
        <dd>Check if there is a connection to a BED database</dd>
      </dl><dl><dt>
          
          <code><a href="cleanDubiousXRef.html">cleanDubiousXRef()</a></code> 
        </dt>
        <dd>Identify and remove dubious cross-references</dd>
      </dl><dl><dt>
          
          <code><a href="clearBedCache.html">clearBedCache()</a></code> 
        </dt>
        <dd>Clear the BED cache SQLite database</dd>
      </dl><dl><dt>
          
          <code><a href="compareBedInstances.html">compareBedInstances()</a></code> 
        </dt>
        <dd>Compare 2 BED database instances</dd>
      </dl><dl><dt>
          
          <code><a href="connectToBed.html">connectToBed()</a></code> 
        </dt>
        <dd>Connect to a neo4j BED database</dd>
      </dl><dl><dt>
          
          <code><a href="convBeIdLists.html">convBeIdLists()</a></code> 
        </dt>
        <dd>Converts lists of BE IDs</dd>
      </dl><dl><dt>
          
          <code><a href="convBeIds.html">convBeIds()</a></code> 
        </dt>
        <dd>Converts BE IDs</dd>
      </dl><dl><dt>
          
          <code><a href="convDfBeIds.html">convDfBeIds()</a></code> 
        </dt>
        <dd>Add BE ID conversion to a data frame</dd>
      </dl><dl><dt>
          
          <code><a href="dumpEnsCore.html">dumpEnsCore()</a></code> 
        </dt>
        <dd>Feeding BED: Dump table from the Ensembl core database</dd>
      </dl><dl><dt>
          
          <code><a href="dumpNcbiDb.html">dumpNcbiDb()</a></code> 
        </dt>
        <dd>Feeding BED: Dump tables from the NCBI gene DATA</dd>
      </dl><dl><dt>
          
          <code><a href="dumpNcbiTax.html">dumpNcbiTax()</a></code> 
        </dt>
        <dd>Feeding BED: Dump tables with taxonomic information from NCBI</dd>
      </dl><dl><dt>
          
          <code><a href="dumpUniprotDb.html">dumpUniprotDb()</a></code> 
        </dt>
        <dd>Feeding BED: Dump and preprocess flat dat files fro Uniprot</dd>
      </dl><dl><dt>
          
          <code><a href="exploreBe.html">exploreBe()</a></code> 
        </dt>
        <dd>Explore BE identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="exploreConvPath.html">exploreConvPath()</a></code> 
        </dt>
        <dd>Explore the shortest convertion path between two identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="filterByBEID.html">filterByBEID()</a></code> 
        </dt>
        <dd>Filter an object to keep only a set of BEIDs</dd>
      </dl><dl><dt>
          
          <code><a href="findBe.html">findBe()</a></code> 
        </dt>
        <dd>Find Biological Entity</dd>
      </dl><dl><dt>
          
          <code><a href="findBeids.html">findBeids()</a></code> 
        </dt>
        <dd>Find Biological Entity identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="firstCommonUpstreamBe.html">firstCommonUpstreamBe()</a></code> 
        </dt>
        <dd>First common upstream BE</dd>
      </dl><dl><dt>
          
          <code><a href="focusOnScope.BEIDList.html">focusOnScope(<i>&lt;BEIDList&gt;</i>)</a></code> 
        </dt>
        <dd>Convert a BEIDList object in a specific identifier (BEID) scope</dd>
      </dl><dl><dt>
          
          <code><a href="focusOnScope.html">focusOnScope()</a></code> 
        </dt>
        <dd>Focus a BE related object on a specific identifier (BEID) scope</dd>
      </dl><dl><dt>
          
          <code><a href="forgetBedConnection.html">forgetBedConnection()</a></code> 
        </dt>
        <dd>Forget a BED connection</dd>
      </dl><dl><dt>
          
          <code><a href="genBePath.html">genBePath()</a></code> 
        </dt>
        <dd>Construct CQL sub-query to map 2 biological entity</dd>
      </dl><dl><dt>
          
          <code><a href="genProbePath.html">genProbePath()</a></code> 
        </dt>
        <dd>Identify the biological entity (BE) targeted by probes and
construct the CQL sub-query to map probes to the BE</dd>
      </dl><dl><dt>
          
          <code><a href="geneIDsToAllScopes.html">geneIDsToAllScopes()</a></code> 
        </dt>
        <dd>Find all GeneID, ObjectID, TranscriptID, PeptideID and ProbeID corresponding to a Gene in any organism</dd>
      </dl><dl><dt>
          
          <code><a href="getAllBeIdSources.html">getAllBeIdSources()</a></code> 
        </dt>
        <dd>List all the source databases of BE identifiers whatever the BE type</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdConvTable.html">getBeIdConvTable()</a></code> 
        </dt>
        <dd>Get a conversion table between biological entity (BE) identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdDescription.html">getBeIdDescription()</a></code> 
        </dt>
        <dd>Get description of Biological Entity identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdNameTable.html">getBeIdNameTable()</a></code> 
        </dt>
        <dd>Get a table of biological entity (BE) identifiers and names</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdNames.html">getBeIdNames()</a></code> 
        </dt>
        <dd>Get names of Biological Entity identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdSymbolTable.html">getBeIdSymbolTable()</a></code> 
        </dt>
        <dd>Get a table of biological entity (BE) identifiers and symbols</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdSymbols.html">getBeIdSymbols()</a></code> 
        </dt>
        <dd>Get symbols of Biological Entity identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIdURL.html">getBeIdURL()</a></code> 
        </dt>
        <dd>Get reference URLs for BE IDs</dd>
      </dl><dl><dt>
          
          <code><a href="getBeIds.html">getBeIds()</a></code> 
        </dt>
        <dd>Get biological entities identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getDirectOrigin.html">getDirectOrigin()</a></code> 
        </dt>
        <dd>Get the direct origin of BE identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getDirectProduct.html">getDirectProduct()</a></code> 
        </dt>
        <dd>Get the direct product of BE identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getEnsemblGeneIds.html">getEnsemblGeneIds()</a></code> 
        </dt>
        <dd>Feeding BED: Download Ensembl DB and load gene information in BED</dd>
      </dl><dl><dt>
          
          <code><a href="getEnsemblPeptideIds.html">getEnsemblPeptideIds()</a></code> 
        </dt>
        <dd>Feeding BED: Download Ensembl DB and load peptide information in BED</dd>
      </dl><dl><dt>
          
          <code><a href="getEnsemblTranscriptIds.html">getEnsemblTranscriptIds()</a></code> 
        </dt>
        <dd>Feeding BED: Download Ensembl DB and load transcript information in BED</dd>
      </dl><dl><dt>
          
          <code><a href="getGeneDescription.html">getGeneDescription()</a></code> 
        </dt>
        <dd>Get description of genes corresponding to Biological Entity identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="getHomTable.html">getHomTable()</a></code> 
        </dt>
        <dd>Get gene homologs between 2 organisms</dd>
      </dl><dl><dt>
          
          <code><a href="getNcbiGeneTransPep.html">getNcbiGeneTransPep()</a></code> 
        </dt>
        <dd>Feeding BED: Download NCBI gene DATA and load gene, transcript and peptide
information in BED</dd>
      </dl><dl><dt>
          
          <code><a href="getOrgNames.html">getOrgNames()</a></code> 
        </dt>
        <dd>Get organism names from taxonomy IDs</dd>
      </dl><dl><dt>
          
          <code><a href="getRelevantIds.html">getRelevantIds()</a></code> 
        </dt>
        <dd>Get relevant IDs for a formerly identified BE in a context of interest</dd>
      </dl><dl><dt>
          
          <code><a href="getTargetedBe.html">getTargetedBe()</a></code> 
        </dt>
        <dd>Identify the biological entity (BE) targeted by probes</dd>
      </dl><dl><dt>
          
          <code><a href="getTaxId.html">getTaxId()</a></code> 
        </dt>
        <dd>Get taxonomy ID of an organism name</dd>
      </dl><dl><dt>
          
          <code><a href="getUniprot.html">getUniprot()</a></code> 
        </dt>
        <dd>Feeding BED: Download Uniprot information in BED</dd>
      </dl><dl><dt>
          
          <code><a href="guessIdScope.html">guessIdScope()</a></code> <code><a href="guessIdScope.html">guessIdOrigin()</a></code> 
        </dt>
        <dd>Guess biological entity (BE), database source and organism of a vector
of identifiers.</dd>
      </dl><dl><dt>
          
          <code><a href="identicalScopes.html">identicalScopes()</a></code> 
        </dt>
        <dd>Check if two objects have the same BEID scope</dd>
      </dl><dl><dt>
          
          <code><a href="is.BEIDList.html">is.BEIDList()</a></code> 
        </dt>
        <dd>Check if the provided object is a BEIDList</dd>
      </dl><dl><dt>
          
          <code><a href="largestBeSource.html">largestBeSource()</a></code> 
        </dt>
        <dd>Autoselect source of biological entity identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="listBe.html">listBe()</a></code> 
        </dt>
        <dd>Lists all the biological entities (BE) available in the BED database</dd>
      </dl><dl><dt>
          
          <code><a href="listBeIdSources.html">listBeIdSources()</a></code> 
        </dt>
        <dd>Lists all the databases taken into account in the BED database
for a biological entity (BE)</dd>
      </dl><dl><dt>
          
          <code><a href="listDBAttributes.html">listDBAttributes()</a></code> 
        </dt>
        <dd>List all attributes provided by a BEDB</dd>
      </dl><dl><dt>
          
          <code><a href="listOrganisms.html">listOrganisms()</a></code> 
        </dt>
        <dd>Lists all the organisms available in the BED database</dd>
      </dl><dl><dt>
          
          <code><a href="listPlatforms.html">listPlatforms()</a></code> 
        </dt>
        <dd>Lists all the probe platforms available in the BED database</dd>
      </dl><dl><dt>
          
          <code><a href="loadBE.html">loadBE()</a></code> 
        </dt>
        <dd>Feeding BED: Load biological entities in BED</dd>
      </dl><dl><dt>
          
          <code><a href="loadBENames.html">loadBENames()</a></code> 
        </dt>
        <dd>Feeding BED: Load names associated to BEIDs</dd>
      </dl><dl><dt>
          
          <code><a href="loadBESymbols.html">loadBESymbols()</a></code> 
        </dt>
        <dd>Feeding BED: Load symbols associated to BEIDs</dd>
      </dl><dl><dt>
          
          <code><a href="loadBEVersion.html">loadBEVersion()</a></code> 
        </dt>
        <dd>Feeding BED: Load biological entities in BED with information about
DB version</dd>
      </dl><dl><dt>
          
          <code><a href="loadBeAttribute.html">loadBeAttribute()</a></code> 
        </dt>
        <dd>Feeding BED: Load attributes for biological entities in BED</dd>
      </dl><dl><dt>
          
          <code><a href="loadBedModel.html">loadBedModel()</a></code> 
        </dt>
        <dd>Feeding BED: Load BED data model in neo4j</dd>
      </dl><dl><dt>
          
          <code><a href="loadBedOtherIndexes.html">loadBedOtherIndexes()</a></code> 
        </dt>
        <dd>Feeding BED: Load additional indexes in neo4j</dd>
      </dl><dl><dt>
          
          <code><a href="loadBedResult.html">loadBedResult()</a></code> 
        </dt>
        <dd>Get a BED query result from cache</dd>
      </dl><dl><dt>
          
          <code><a href="loadCodesFor.html">loadCodesFor()</a></code> 
        </dt>
        <dd>Feeding BED: Load correspondance between genes and objects as coding events</dd>
      </dl><dl><dt>
          
          <code><a href="loadCorrespondsTo.html">loadCorrespondsTo()</a></code> 
        </dt>
        <dd>Feeding BED: Load correspondances between BE IDs</dd>
      </dl><dl><dt>
          
          <code><a href="loadHistory.html">loadHistory()</a></code> 
        </dt>
        <dd>Feeding BED: Load history of BEIDs</dd>
      </dl><dl><dt>
          
          <code><a href="loadIsAssociatedTo.html">loadIsAssociatedTo()</a></code> 
        </dt>
        <dd>Feeding BED: Load BE ID associations</dd>
      </dl><dl><dt>
          
          <code><a href="loadIsExpressedAs.html">loadIsExpressedAs()</a></code> 
        </dt>
        <dd>Feeding BED: Load correspondance between genes and transcripts as
expression events</dd>
      </dl><dl><dt>
          
          <code><a href="loadIsHomologOf.html">loadIsHomologOf()</a></code> 
        </dt>
        <dd>Feeding BED: Load homology between BE IDs</dd>
      </dl><dl><dt>
          
          <code><a href="loadIsTranslatedIn.html">loadIsTranslatedIn()</a></code> 
        </dt>
        <dd>Feeding BED: Load correspondance between transcripts and peptides as
translation events</dd>
      </dl><dl><dt>
          
          <code><a href="loadLuceneIndexes.html">loadLuceneIndexes()</a></code> 
        </dt>
        <dd>Feeding BED: Create Lucene indexes in neo4j</dd>
      </dl><dl><dt>
          
          <code><a href="loadNCBIEntrezGOFunctions.html">loadNCBIEntrezGOFunctions()</a></code> 
        </dt>
        <dd>Feeding BED: Load in BED GO functions associated to Entrez gene IDs
from NCBI</dd>
      </dl><dl><dt>
          
          <code><a href="loadNcbiTax.html">loadNcbiTax()</a></code> 
        </dt>
        <dd>Feeding BED: Load taxonomic information from NCBI</dd>
      </dl><dl><dt>
          
          <code><a href="loadOrganisms.html">loadOrganisms()</a></code> 
        </dt>
        <dd>Feeding BED: Load organisms in BED</dd>
      </dl><dl><dt>
          
          <code><a href="loadPlf.html">loadPlf()</a></code> 
        </dt>
        <dd>Feeding BED: Load a probes platform</dd>
      </dl><dl><dt>
          
          <code><a href="loadProbes.html">loadProbes()</a></code> 
        </dt>
        <dd>Feeding BED: Load probes targeting BE IDs</dd>
      </dl><dl><dt>
          
          <code><a href="lsBedCache.html">lsBedCache()</a></code> 
        </dt>
        <dd>List all the BED queries in cache and the total size of the cache</dd>
      </dl><dl><dt>
          
          <code><a href="lsBedConnections.html">lsBedConnections()</a></code> 
        </dt>
        <dd>List all registered BED connection</dd>
      </dl><dl><dt>
          
          <code><a href="metadata-set.html">`metadata&lt;-`()</a></code> 
        </dt>
        <dd>Set object metadata</dd>
      </dl><dl><dt>
          
          <code><a href="metadata.html">metadata()</a></code> 
        </dt>
        <dd>Get object metadata</dd>
      </dl><dl><dt>
          
          <code><a href="registerBEDB.html">registerBEDB()</a></code> 
        </dt>
        <dd>Feeding BED: Register a database of biological entities in BED DB</dd>
      </dl><dl><dt>
          
          <code><a href="scope.html">scope()</a></code> 
        </dt>
        <dd>Get the BEID scope of an object</dd>
      </dl><dl><dt>
          
          <code><a href="scopes.html">scopes()</a></code> 
        </dt>
        <dd>Get the BEID scopes of an object</dd>
      </dl><dl><dt>
          
          <code><a href="searchBeid.html">searchBeid()</a></code> 
        </dt>
        <dd>Search a BEID</dd>
      </dl><dl><dt>
          
          <code><a href="searchId.html">searchId()</a></code> 
        </dt>
        <dd>Search identifier, symbol or name information</dd>
      </dl><dl><dt>
          
          <code><a href="setBedVersion.html">setBedVersion()</a></code> 
        </dt>
        <dd>Feeding BED: Set the BED version</dd>
      </dl><dl><dt>
          
          <code><a href="showBedDataModel.html">showBedDataModel()</a></code> 
        </dt>
        <dd>Show the data model of BED</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/patzaw" class="external-link">Patrice Godard</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

